<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>The structure of Pangraph output file · PanGraph.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PanGraph.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorial</span><ul><li><a class="tocitem" href="../tutorial_1/">Building a pangraph</a></li><li class="is-active"><a class="tocitem" href>The structure of Pangraph output file</a><ul class="internal"><li><a class="tocitem" href="#The-structure-of-pangraph.json"><span>The structure of <code>pangraph.json</code></span></a></li><li><a class="tocitem" href="#Blocks"><span>Blocks</span></a></li><li><a class="tocitem" href="#How-alignments-are-encoded"><span>How alignments are encoded</span></a></li><li><a class="tocitem" href="#Paths"><span>Paths</span></a></li><li><a class="tocitem" href="#A-look-at-the-length-and-frequency-of-blocks"><span>A look at the length and frequency of blocks</span></a></li></ul></li><li><a class="tocitem" href="../tutorial_3/">Polishing the pangraph and exploring alignments</a></li><li><a class="tocitem" href="../tutorial_4/">Projecting the graph on a subset of strains.</a></li><li><a class="tocitem" href="../tutorial_5/">Example application: plasmid rearrangements</a></li><li><a class="tocitem" href="../tutorial_6/">Example application: antibiotic resistance gene neighbourhoods</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/pangraph/">PanGraph globals</a></li><li><a class="tocitem" href="../../lib/align/">Alignment</a></li><li><a class="tocitem" href="../../lib/block/">Blocks</a></li><li><a class="tocitem" href="../../lib/edge/">Edges</a></li><li><a class="tocitem" href="../../lib/graph/">Graphs</a></li><li><a class="tocitem" href="../../lib/gfa/">GFA</a></li><li><a class="tocitem" href="../../lib/mash/">Mash Implementation</a></li><li><a class="tocitem" href="../../lib/minimap/">Minimap2 Wrapper</a></li><li><a class="tocitem" href="../../lib/mmseqs/">MMseqs Wrapper</a></li><li><a class="tocitem" href="../../lib/node/">Nodes</a></li><li><a class="tocitem" href="../../lib/path/">Paths</a></li><li><a class="tocitem" href="../../lib/simulate/">Simulation</a></li><li><a class="tocitem" href="../../lib/utility/">Utility</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/build/">Build</a></li><li><a class="tocitem" href="../../cli/export/">Export</a></li><li><a class="tocitem" href="../../cli/generate/">Generate</a></li><li><a class="tocitem" href="../../cli/marginalize/">Marginalize</a></li><li><a class="tocitem" href="../../cli/polish/">Polish</a></li><li><a class="tocitem" href="../../cli/version/">Version</a></li></ul></li><li><span class="tocitem">Development</span><ul><li><a class="tocitem" href="../../dev/building-docker/">Building Docker image</a></li><li><a class="tocitem" href="../../dev/building-documentation/">Updating the documentation</a></li><li><a class="tocitem" href="../../dev/releasing/">Releasing</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorial</a></li><li class="is-active"><a href>The structure of Pangraph output file</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>The structure of Pangraph output file</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/neherlab/pangraph/blob/master/docs/src/tutorials/tutorial_2.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="The-structure-of-Pangraph-output-file"><a class="docs-heading-anchor" href="#The-structure-of-Pangraph-output-file">The structure of Pangraph output file</a><a id="The-structure-of-Pangraph-output-file-1"></a><a class="docs-heading-anchor-permalink" href="#The-structure-of-Pangraph-output-file" title="Permalink"></a></h1><p>In this second part of the tutorial we will explore in more detail the content of the <code>json</code> output file produced by the <code>build</code> command. As an example, we will use snippets from the <code>ecoli_pangraph.json</code> file that was produced in the previous section of the tutorial.</p><h2 id="The-structure-of-pangraph.json"><a class="docs-heading-anchor" href="#The-structure-of-pangraph.json">The structure of <code>pangraph.json</code></a><a id="The-structure-of-pangraph.json-1"></a><a class="docs-heading-anchor-permalink" href="#The-structure-of-pangraph.json" title="Permalink"></a></h2><p>As discussed in the previous tutorial section, the two main entries of pangraph output file are <code>paths</code> and <code>blocks</code>.</p><ul><li>each entry in the <code>paths</code> list encodes one of the nucleotide sequences that were given as input to the <code>build</code> command, represented as a list of blocks.</li><li>each entry in the <code>blocks</code> list represents an alignable set of homologous sequences. A block contains the consensus of all of these sequences, together with information to reconstruct the full alignment.</li></ul><p>We will explore each of these two categories separately.</p><h2 id="Blocks"><a class="docs-heading-anchor" href="#Blocks">Blocks</a><a id="Blocks-1"></a><a class="docs-heading-anchor-permalink" href="#Blocks" title="Permalink"></a></h2><p>Here is an example of an entry of the <code>blocks</code> list, from the <code>ecoli_pangraph.json</code> file.</p><pre><code class="language-json hljs">{
    &quot;id&quot;: &quot;TMEPNAOFAP&quot;,
    &quot;sequence&quot;: &quot;CGGGAAGGTTCTGATGCGTCCGTGTTAAACTAAGAGAATCTATCT...&quot;,
    &quot;gaps&quot;: { ... },
    &quot;mutate&quot;: [ ... ],
    &quot;insert&quot;: [ ... ],
    &quot;delete&quot;: [ ... ],
    &quot;positions&quot;: [
        [
            { &quot;name&quot;: &quot;NZ_CP019944&quot;, &quot;number&quot;: 1, &quot;strand&quot;: true },
            [ 356656, 359732 ]
        ],
        [
            { &quot;name&quot;: &quot;NC_013361&quot;, &quot;number&quot;: 1, &quot;strand&quot;: false},
            [ 4496907, 4500032 ]
        ],
        [
            { &quot;name&quot;: &quot;NZ_CP015912&quot;, &quot;number&quot;: 1, &quot;strand&quot;: false},
            [ 2485802, 2488871 ]
        ],
        ...
    ]
},</code></pre><p>The two main properties of a block are its unique <code>id</code> (10-letters alphabetic sequence randomly assigned when generating the pangraph), and the consensus <code>sequence</code>.</p><p>In addition to this, a block object also stores information on the full alignment of all the different occurrences of the block, and their position on the original input sequences. As explained in the <a href="../../#Introduction">Introduction</a>, we refer to a particular occurrence of a block in a sequence as a <strong>Node</strong>. Intuitively, they can be thought of as a single entry in the sequence alignment of a block.</p><p>Each node of a block can be uniquely identified by its &quot;node-id&quot;, which is an object composed of three entries:</p><pre><code class="language-json hljs">{ &quot;name&quot;: &quot;NZ_CP019944&quot;, &quot;number&quot;: 1, &quot;strand&quot;: true }</code></pre><ul><li>a <code>name</code>, indicating on which of the input sequences the node is found.</li><li>a <code>number</code>, indicating whether this is the first, second, third... occurrence of the block on a given input sequence. This is important to distinguish different nodes of a duplicated blocks, when they occurr in the same input sequence.</li><li>a <code>strand</code> boolean value, indicating wheter the node is found on the forward (<code>true</code>) or reverse (<code>false</code>) strand.</li></ul><p>The &quot;node-id&quot; is used for example in the <code>position</code> field of a block. This field contains information on the location of nodes on the input sequences. It consists of a list of pairs. The first entry of the pair is the node-id, indicating of which occurrence of the block the position refers to. The second entry is a pair of numbers indicating the position of the beginning and end of the node on the input sequence.</p><p>In the example above, the first entry of <code>positions</code> indicates that block <code>&quot;TMEPNAOFAP&quot;</code> is found on the forward strand of the chromosome labeled <code>NZ_CP019944</code>, at position <code>356656</code> to <code>359732</code>. Positions are always in 1-based numbering and based on the forward strand (with beginning &lt; end). The only exception is when a block wraps around the end of a circular sequence. In this case the node start position (close to the end of the genome) is higher than the node end position (close to the beginning of the genome).</p><p>The fields <code>gaps</code>, <code>mutate</code>, <code>insert</code> and <code>delete</code> contain information to reconstruct the block alignment. They were left out from this example for simplicity, and are discussed in the next subsection.</p><h2 id="How-alignments-are-encoded"><a class="docs-heading-anchor" href="#How-alignments-are-encoded">How alignments are encoded</a><a id="How-alignments-are-encoded-1"></a><a class="docs-heading-anchor-permalink" href="#How-alignments-are-encoded" title="Permalink"></a></h2><p>A block object contains the information on the alignment of all the homologous sequences belonging to the block. This information is stored in compressed format in the <code>gaps</code>, <code>mutate</code>, <code>insert</code> and <code>delete</code> fields. In practice the alignment of each node sequence can be obtained from the consensus by adding the gaps, single-nucleotide mutations, and insertions / deletions contained in these fields.</p><p>Here we briefly describe each entry, and then schematically show how these can be combined to obtain the alignment sequence of a node.</p><div class="admonition is-info"><header class="admonition-header">reconstructing the alignments</header><div class="admonition-body"><p>Rather than providing explicit instructions on how to build the alignments, the aim of this part of the tutorial is to show how the information is organized in the pangraph output format. For a quick way to obtain block alignments, have a look at the next tutorial section <a href="../tutorial_3/#Polishing-the-pangraph-and-exploring-alignments">Polishing the pangraph and exploring alignments</a>.</p></div></div><h3 id="Gaps"><a class="docs-heading-anchor" href="#Gaps">Gaps</a><a id="Gaps-1"></a><a class="docs-heading-anchor-permalink" href="#Gaps" title="Permalink"></a></h3><p>The <code>gaps</code> entry contains a json object that indicates where gaps should be added to the consensus sequence as a first step towards reconstructing the alignment. The first item represents the position of the gap, and the second its length.</p><pre><code class="language-json hljs">&quot;gaps&quot; : {&quot;0&quot;: 27, &quot;53&quot;: 1 }</code></pre><p>In this example the consensus sequence contains two gaps: one is inserted at the beginning of the consensus sequence, with length 27, and one after sequence position 53, with length 1. In some of the sequences running through this block, these gaps are filled with insertions, see below.</p><h3 id="Single-nucleotide-mutations"><a class="docs-heading-anchor" href="#Single-nucleotide-mutations">Single-nucleotide mutations</a><a id="Single-nucleotide-mutations-1"></a><a class="docs-heading-anchor-permalink" href="#Single-nucleotide-mutations" title="Permalink"></a></h3><p>The <code>mutate</code> entry encodes single-nucleotide mutations. It contains a list of pairs. Each pair is relative to a particular node (i.e. one line in the alignment) and the first item of the pair is the node-id described above. The second entry is a list of mutations (potentially empty). Each mutation is specified by its position on the original sequence, and the nucleotide that should be substituted to the consensus.</p><pre><code class="language-json hljs">&quot;mutate&quot; : [
    [
        { &quot;name&quot;: &quot;NC_009800&quot;, &quot;number&quot;: 1, &quot;strand&quot;: false },     // node-id
        [ [415,&quot;C&quot;], [1526,&quot;T&quot;], [2827,&quot;G&quot;], ... ]                 // list of mutations
    ],
    [
        { &quot;name&quot;: &quot;NZ_CP011342&quot;, &quot;number&quot;: 1, &quot;strand&quot;: false },   // node-id
        [ [2221, &quot;G&quot;], [415, &quot;C&quot;], [2494, &quot;T&quot;], [1825, &quot;A&quot;], ... ] // list of mutations
    ],
    [
        { &quot;name&quot;: &quot;NZ_CP019944&quot;, &quot;number&quot;: 1, &quot;strand&quot;: true },    // node-id
        [ [2827, &quot;G&quot;], [442, &quot;T&quot;], [2713, &quot;G&quot;], ... ]              // list of mutations
    ],
    ...
],</code></pre><h3 id="Insertions-and-deletions"><a class="docs-heading-anchor" href="#Insertions-and-deletions">Insertions and deletions</a><a id="Insertions-and-deletions-1"></a><a class="docs-heading-anchor-permalink" href="#Insertions-and-deletions" title="Permalink"></a></h3><p>Entries of <code>insert</code> and <code>delete</code> are organized similarly to mutations. They appear in lists of pairs, whose first item is the node-id, and the second is the list of insertions/deletions that one needs to include. This list can potentially be empty.</p><pre><code class="language-json hljs">&quot;insert&quot; : [
    [
        { &quot;name&quot;: &quot;NC_009800&quot;, &quot;number&quot;: 1, &quot;strand&quot;: false },
        []
    ],
    [
        { &quot;name&quot;: &quot;NZ_CP011342&quot;, &quot;number&quot;: 1, &quot;strand&quot;: false },
        []
    ],
    [
        { &quot;name&quot;: &quot;NZ_CP019944&quot;, &quot;number&quot;: 1, &quot;strand&quot;: true },
        [ [ [0,0], &quot;TT&quot;], [ [0,15], &quot;TTCCC&quot;] ]
    ],
    ...
],</code></pre><p>Insertions are always placed inside of gaps. They appear in the form <code>[ [gap-beg, gap-offset], &quot;seq&quot;]</code>. <code>gap-beg</code> indicates the beginning position of the gap, and <code>gap-offset</code> indicates how far from the beginning of this gap the insertion starts. Finally <code>seq</code> is the nucleotide sequence to be inserted.</p><pre><code class="language-json hljs">&quot;delete&quot; : [
    [
        { &quot;name&quot;: &quot;NC_009800&quot;, &quot;number&quot;: 1, &quot;strand&quot;: false },
        [ [ 3071, 56 ] ]
    ],
    [
        { &quot;name&quot;: &quot;NZ_CP011342&quot;, &quot;number&quot;: 1, &quot;strand&quot;: false },
        []
    ],
    [
        { &quot;name&quot;: &quot;NZ_CP019944&quot;, &quot;number&quot;: 1, &quot;strand&quot;: true },
        [ [ 3071, 56 ] ]
    ],
    ...
],</code></pre><p>Deletions are instead in the form <code>[del-beg, del-length]</code> where the first number indicates the position at which the deletion starts (relative to the consensus sequence) and the second is the deletion length.</p><h3 id="From-block-consensus-to-node-alignment"><a class="docs-heading-anchor" href="#From-block-consensus-to-node-alignment">From block consensus to node alignment</a><a id="From-block-consensus-to-node-alignment-1"></a><a class="docs-heading-anchor-permalink" href="#From-block-consensus-to-node-alignment" title="Permalink"></a></h3><p>Below is a schematic summary of how gaps, mutations, insertions and deletions can be combined to go from the block consensus sequence to the alignment sequence of a particular node.</p><p><img src="../../assets/alignment_scheme.png" alt="img"/></p><h2 id="Paths"><a class="docs-heading-anchor" href="#Paths">Paths</a><a id="Paths-1"></a><a class="docs-heading-anchor-permalink" href="#Paths" title="Permalink"></a></h2><p>A path object has the following structure:</p><pre><code class="language-json hljs">{
    &quot;name&quot;: &quot;NZ_CP010242&quot;,
    &quot;offset&quot;: -3044650,
    &quot;circular&quot;: true,
    &quot;position&quot;: [ 1695501, 1696243, 1702666, ... ],
    &quot;blocks&quot;: [
        { &quot;id&quot;: &quot;IUZTZPLBVS&quot;, &quot;name&quot;: &quot;NZ_CP010242&quot;, &quot;number&quot;: 1, &quot;strand&quot;: true },
        { &quot;id&quot;: &quot;DVEJPZTOYJ&quot;, &quot;name&quot;: &quot;NZ_CP010242&quot;, &quot;number&quot;: 1, &quot;strand&quot;: true },
        { &quot;id&quot;: &quot;FUECJYRQOZ&quot;, &quot;name&quot;: &quot;NZ_CP010242&quot;, &quot;number&quot;: 1, &quot;strand&quot;: true },
        ...
    ]
},</code></pre><p>The two main properties of a path are <code>name</code> and <code>blocks</code>. The <code>name</code> of the path indicates which of the input sequences the path represents. <code>blocks</code> contains the ordered list of nodes that make up the path. Each node is identified by the unique block <code>id</code>, and by the entries of the node-id (<code>name</code>, <code>number</code>, <code>strand</code>).</p><p>Here is a complete list containing a description of every entry in the path object:</p><ul><li><code>name</code> : the name of the particular nucleotide sequence to whom the path refers. This name is extracted from the sequence id in the input fasta file.</li><li><code>circular</code> : indicates whether the considered sequence is circular (e.g. plasmid) or not. This is controlled by the <code>--circular</code> option of the <code>build</code> command.</li><li><code>blocks</code> : the ordered list of blocks that make up the path.<ul><li><code>id</code> : the unique random id of the block, assigned when building the graph.</li><li><code>name</code>, <code>number</code> <code>strand</code> : entries of the node-id, used to identify which particular instance of the block is part of the path. As a reminder, <code>name</code> is the id of the input sequence, <code>number</code> indicates which occurrence of the same block is considered (useful for duplicated blocks) and <code>strand</code> indicates whether the sequence is found on the forward or reverse strand.</li></ul></li><li><code>position</code> : an ordered list of positions, corresponding to the beginning positions of each node in the path. If N is the number of nodes this list has N+1 entries. The last entry is the position of the right edge of the last block in the path.</li><li><code>offset</code> : indicates the distance between the beginning of the input sequence, and the beginning of the path (i.e. the beginning of the first node of the path, block <code>IUZTZPLBVS</code> in the example above).</li></ul><h2 id="A-look-at-the-length-and-frequency-of-blocks"><a class="docs-heading-anchor" href="#A-look-at-the-length-and-frequency-of-blocks">A look at the length and frequency of blocks</a><a id="A-look-at-the-length-and-frequency-of-blocks-1"></a><a class="docs-heading-anchor-permalink" href="#A-look-at-the-length-and-frequency-of-blocks" title="Permalink"></a></h2><p>Having completed this part of the tutorial, it is now possible to access directly the rich information contained in the pangraph output format.</p><p>As a simple example, we take the <code>ecoli_pangraph.json</code> file and extract from each block the length of its consensus sequence, and the number of unique E.Coli strains that contain the block. We then plot the cumulative distribution of block lengths, and the frequency distribution of blocks (weighted by length).</p><p><img src="../../assets/ecoli_blockstats.png" alt="img"/></p><p>The blocks present in this pangraph have widely varying size, with some blocks being only some hundreds of nucleotides long, and others spanning tens of kbps.<sup class="footnote-reference"><a id="citeref-1" href="#footnote-1">[1]</a></sup></p><p>The block frequency distribution shows a typical bimodal pattern, with an abundance of &quot;core&quot; blocks (blocks that are present in all the 10 considered chromosomes, cumulative length of more than 3.5 Mbps) and rare blocks present in only one strain (cumulative length of more than 2.5 Mbps). This graph can be reproduced by this python snippet:</p><pre><code class="language-python hljs">import json
import matplotlib.pyplot as plt

with open(&#39;ecoli_pangraph.json&#39;) as fh:
    G = json.load(fh)

blocks = [{&#39;length&#39;:len(x[&#39;sequence&#39;]), &#39;depth&#39;:len(x[&#39;positions&#39;])} for x in G[&#39;blocks&#39;]]
plt.hist([x[&#39;depth&#39;] for x in blocks], weights=[x[&#39;length&#39;] for x in blocks], bins=range(1,11))</code></pre><section class="footnotes is-size-7"><ul><li class="footnote" id="footnote-1"><a class="tag is-link" href="#citeref-1">1</a>The almost complete absence of blocks shorter than 100 bps is due to the default value of the <code>--len</code> option in the <code>build</code> command, which is set to 100. See <a href="../../cli/build/#Build">Build</a>.</li></ul></section></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../tutorial_1/">« Building a pangraph</a><a class="docs-footer-nextpage" href="../tutorial_3/">Polishing the pangraph and exploring alignments »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 21 November 2023 09:43">Tuesday 21 November 2023</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
