<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Build · PanGraph.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PanGraph.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorial</span><ul><li><a class="tocitem" href="../../tutorials/tutorial_1/">Building a pangraph</a></li><li><a class="tocitem" href="../../tutorials/tutorial_2/">The structure of Pangraph output file</a></li><li><a class="tocitem" href="../../tutorials/tutorial_3/">Polishing the pangraph and exploring alignments</a></li><li><a class="tocitem" href="../../tutorials/tutorial_4/">Projecting the graph on a subset of strains.</a></li><li><a class="tocitem" href="../../tutorials/tutorial_5/">Example application: plasmid rearrangements</a></li><li><a class="tocitem" href="../../tutorials/tutorial_6/">Example application: antibiotic resistance gene neighbourhoods</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/pangraph/">PanGraph globals</a></li><li><a class="tocitem" href="../../lib/align/">Alignment</a></li><li><a class="tocitem" href="../../lib/block/">Blocks</a></li><li><a class="tocitem" href="../../lib/edge/">Edges</a></li><li><a class="tocitem" href="../../lib/graph/">Graphs</a></li><li><a class="tocitem" href="../../lib/gfa/">GFA</a></li><li><a class="tocitem" href="../../lib/mash/">Mash Implementation</a></li><li><a class="tocitem" href="../../lib/minimap/">Minimap2 Wrapper</a></li><li><a class="tocitem" href="../../lib/mmseqs/">MMseqs Wrapper</a></li><li><a class="tocitem" href="../../lib/node/">Nodes</a></li><li><a class="tocitem" href="../../lib/path/">Paths</a></li><li><a class="tocitem" href="../../lib/simulate/">Simulation</a></li><li><a class="tocitem" href="../../lib/utility/">Utility</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li class="is-active"><a class="tocitem" href>Build</a><ul class="internal"><li><a class="tocitem" href="#Description"><span>Description</span></a></li><li><a class="tocitem" href="#Options"><span>Options</span></a></li><li><a class="tocitem" href="#Arguments"><span>Arguments</span></a></li><li><a class="tocitem" href="#Output"><span>Output</span></a></li></ul></li><li><a class="tocitem" href="../export/">Export</a></li><li><a class="tocitem" href="../generate/">Generate</a></li><li><a class="tocitem" href="../marginalize/">Marginalize</a></li><li><a class="tocitem" href="../polish/">Polish</a></li><li><a class="tocitem" href="../version/">Version</a></li></ul></li><li><span class="tocitem">Development</span><ul><li><a class="tocitem" href="../../dev/building-docker/">Building Docker image</a></li><li><a class="tocitem" href="../../dev/building-documentation/">Updating the documentation</a></li><li><a class="tocitem" href="../../dev/releasing/">Releasing</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Command Line</a></li><li class="is-active"><a href>Build</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Build</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/neherlab/pangraph/blob/master/docs/src/cli/build.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Build"><a class="docs-heading-anchor" href="#Build">Build</a><a id="Build-1"></a><a class="docs-heading-anchor-permalink" href="#Build" title="Permalink"></a></h1><h2 id="Description"><a class="docs-heading-anchor" href="#Description">Description</a><a id="Description-1"></a><a class="docs-heading-anchor-permalink" href="#Description" title="Permalink"></a></h2><p>Build a multiple sequence alignment pangraph.</p><h2 id="Options"><a class="docs-heading-anchor" href="#Options">Options</a><a id="Options-1"></a><a class="docs-heading-anchor-permalink" href="#Options" title="Permalink"></a></h2><table><tr><th style="text-align: left">Name</th><th style="text-align: left">Type</th><th style="text-align: left">Short Flag</th><th style="text-align: left">Long Flag</th><th style="text-align: left">Description</th></tr><tr><td style="text-align: left">minimum length</td><td style="text-align: left">Integer</td><td style="text-align: left">l</td><td style="text-align: left">len</td><td style="text-align: left">minimum block size for alignment graph (in nucleotides)</td></tr><tr><td style="text-align: left">block junction cost</td><td style="text-align: left">Float</td><td style="text-align: left">a</td><td style="text-align: left">alpha</td><td style="text-align: left">energy cost for introducing block partitions due to alignment merger</td></tr><tr><td style="text-align: left">block diversity cost</td><td style="text-align: left">Float</td><td style="text-align: left">b</td><td style="text-align: left">beta</td><td style="text-align: left">energy cost for interblock diversity due to alignment merger</td></tr><tr><td style="text-align: left">circular genomes</td><td style="text-align: left">Boolean</td><td style="text-align: left">c</td><td style="text-align: left">circular</td><td style="text-align: left">toggle if input genomes are circular</td></tr><tr><td style="text-align: left">pairwise sensitivity</td><td style="text-align: left">String</td><td style="text-align: left">s</td><td style="text-align: left">sensitivity</td><td style="text-align: left">controls the pairwise genome alignment sensitivity of minimap 2. Currently only accepts &quot;5&quot;, &quot;10&quot; or &quot;20&quot;</td></tr><tr><td style="text-align: left">maximum self-maps</td><td style="text-align: left">Integer</td><td style="text-align: left">x</td><td style="text-align: left">max-self-map</td><td style="text-align: left">maximum number of iterations to perform block self maps per pairwise graph merger</td></tr><tr><td style="text-align: left">enforce uppercase</td><td style="text-align: left">Boolean</td><td style="text-align: left">u</td><td style="text-align: left">upper-case</td><td style="text-align: left">toggle to force genomes to uppercase characters</td></tr><tr><td style="text-align: left">distance calculator</td><td style="text-align: left">String</td><td style="text-align: left">d</td><td style="text-align: left">distance-backend</td><td style="text-align: left">only accepts &quot;native&quot; or &quot;mash&quot;</td></tr><tr><td style="text-align: left">alignment kernel</td><td style="text-align: left">String</td><td style="text-align: left">k</td><td style="text-align: left">alignment-kernel</td><td style="text-align: left">only accepts &quot;minimap2&quot; or &quot;mmseqs&quot;</td></tr><tr><td style="text-align: left">kmer length (mmseqs)</td><td style="text-align: left">Integer</td><td style="text-align: left">K</td><td style="text-align: left">kmer-length</td><td style="text-align: left">kmer length, only used for mmseqs2 alignment kernel. If not specified will use mmseqs default.</td></tr><tr><td style="text-align: left">consistency check</td><td style="text-align: left">Boolean</td><td style="text-align: left">t</td><td style="text-align: left">test</td><td style="text-align: left">toggle to activate consistency check: verifies that input genomes can be exactly reconstructed from the graph</td></tr><tr><td style="text-align: left">random seed</td><td style="text-align: left">Int</td><td style="text-align: left">r</td><td style="text-align: left">random-seed</td><td style="text-align: left">random seed for pangraph construction.</td></tr></table><h2 id="Arguments"><a class="docs-heading-anchor" href="#Arguments">Arguments</a><a id="Arguments-1"></a><a class="docs-heading-anchor-permalink" href="#Arguments" title="Permalink"></a></h2><p>Expects one or more fasta files. Multiple records within one file are treated as separate genomes Fasta files can be optionally gzipped.</p><h2 id="Output"><a class="docs-heading-anchor" href="#Output">Output</a><a id="Output-1"></a><a class="docs-heading-anchor-permalink" href="#Output" title="Permalink"></a></h2><p>Prints the constructed pangraph as a JSON to <em>stdout</em>.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../lib/utility/">« Utility</a><a class="docs-footer-nextpage" href="../export/">Export »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Wednesday 12 July 2023 14:52">Wednesday 12 July 2023</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
