<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Graphs · PanGraph.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><script src="../../../copy.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PanGraph.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorial</span><ul><li><a class="tocitem" href="../../tutorials/tutorial_1/">Building a pangraph</a></li><li><a class="tocitem" href="../../tutorials/tutorial_2/">The structure of Pangraph output file</a></li><li><a class="tocitem" href="../../tutorials/tutorial_3/">Polishing the pangraph and exploring alignments</a></li><li><a class="tocitem" href="../../tutorials/tutorial_4/">Projecting the graph on a subset of strains.</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../align/">Alignment</a></li><li><a class="tocitem" href="../block/">Blocks</a></li><li><a class="tocitem" href="../edge/">Edges</a></li><li class="is-active"><a class="tocitem" href>Graphs</a><ul class="internal"><li><a class="tocitem" href="#Types"><span>Types</span></a></li><li><a class="tocitem" href="#Functions"><span>Functions</span></a></li></ul></li><li><a class="tocitem" href="../gfa/">GFA</a></li><li><a class="tocitem" href="../mash/">Mash Implementation</a></li><li><a class="tocitem" href="../minimap/">Minimap2 Wrapper</a></li><li><a class="tocitem" href="../node/">Nodes</a></li><li><a class="tocitem" href="../path/">Paths</a></li><li><a class="tocitem" href="../simulate/">Simulation</a></li><li><a class="tocitem" href="../utility/">Utility</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/build/">Build</a></li><li><a class="tocitem" href="../../cli/export/">Export</a></li><li><a class="tocitem" href="../../cli/generate/">Generate</a></li><li><a class="tocitem" href="../../cli/marginalize/">Marginalize</a></li><li><a class="tocitem" href="../../cli/polish/">Polish</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>Graphs</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Graphs</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/neherlab/pangraph/blob/master/docs/src/lib/graph.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Graphs"><a class="docs-heading-anchor" href="#Graphs">Graphs</a><a id="Graphs-1"></a><a class="docs-heading-anchor-permalink" href="#Graphs" title="Permalink"></a></h1><h2 id="Types"><a class="docs-heading-anchor" href="#Types">Types</a><a id="Types-1"></a><a class="docs-heading-anchor-permalink" href="#Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.DelMap" href="#PanGraph.Graphs.DelMap"><code>PanGraph.Graphs.DelMap</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">DelMap = Dict{Int,Int}</code></pre><p>A sparse array of deletion events relative to a consensus. The key is the locus (inclusive) of the deletion; the value is the length.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L65-L70">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Graph" href="#PanGraph.Graphs.Graph"><code>PanGraph.Graphs.Graph</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct Graph
    block    :: Dict{String, Block}
    sequence :: Dict{String, Path}
end</code></pre><p>Representation of a multiple sequence alignment. Alignments of homologous sequences are stored as blocks. A genome is stored as a path, i.e. a list of blocks.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L115-L123">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Graph-Tuple{String, Array{UInt8}}" href="#PanGraph.Graphs.Graph-Tuple{String, Array{UInt8}}"><code>PanGraph.Graphs.Graph</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Graph(name::String, sequence::Array{UInt8}; circular=false)</code></pre><p>Creates a singleton graph from <code>sequence</code>. <code>name</code> is assumed to be a unique identifier. If <code>circular</code> is unspecified, the sequence is assumed to be linear.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L139-L144">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.InsMap" href="#PanGraph.Graphs.InsMap"><code>PanGraph.Graphs.InsMap</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">InsMap = Dict{Tuple{Int,Int},Array{UInt8,1}}</code></pre><p>A sparse array of insertion sequences relative to a consensus. The key is the (locus(after),offset) of the insertion; the value is the sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L58-L63">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.SNPMap" href="#PanGraph.Graphs.SNPMap"><code>PanGraph.Graphs.SNPMap</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">SNPMap = Dict{Int,UInt8}</code></pre><p>A sparse array of single nucleotide polymorphisms relative to a consensus. The key is the locus of the mutation; the value is the modified nucleotide.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L51-L56">source</a></section></article><h2 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.detransitive!-Tuple{PanGraph.Graphs.Graph}" href="#PanGraph.Graphs.detransitive!-Tuple{PanGraph.Graphs.Graph}"><code>PanGraph.Graphs.detransitive!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">detransitive!(G::Graph)</code></pre><p>Find and remove all transitive edges within the given graph. A transitive chain of edges is defined to be unambiguous: all sequences must enter on one edge and leave on another. Thus, this will not perform paralog splitting.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L176-L183">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.finalize!-Tuple{Any}" href="#PanGraph.Graphs.finalize!-Tuple{Any}"><code>PanGraph.Graphs.finalize!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">finalize!(G::Graph)</code></pre><p>Compute the position of the breakpoints for each homologous alignment across all sequences within Graph <code>G</code>. Intended to be ran after multiple sequence alignment is complete</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L605-L610">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.graphs-Tuple{IO}" href="#PanGraph.Graphs.graphs-Tuple{IO}"><code>PanGraph.Graphs.graphs</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">graphs(io::IO; circular=false)</code></pre><p>Parse a fasta file from stream <code>io</code> and return an array of singleton graphs. If circular is unspecified, all genomes are assumed to be linear.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L158-L163">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.keeponly!-Union{Tuple{T}, Tuple{PanGraph.Graphs.Graph, Vararg{T}}} where T&lt;:AbstractString" href="#PanGraph.Graphs.keeponly!-Union{Tuple{T}, Tuple{PanGraph.Graphs.Graph, Vararg{T}}} where T&lt;:AbstractString"><code>PanGraph.Graphs.keeponly!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">keeponly!(G::Graph, names::String...)</code></pre><p>Remove all sequences from graph <code>G</code> that are passed as variadic parameters <code>names</code>. This will marginalize a graph, i.e. return the subgraph that contains only isolates contained in <code>names</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L300-L306">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.marshal_fasta-Tuple{IO, PanGraph.Graphs.Graph}" href="#PanGraph.Graphs.marshal_fasta-Tuple{IO, PanGraph.Graphs.Graph}"><code>PanGraph.Graphs.marshal_fasta</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">marshal_fasta(io::IO, G::Graph; opt=nothing)</code></pre><p>Serialize graph <code>G</code> as a fasta format output stream <code>io</code>. Importantly, this will only serialize the consensus sequences for each block and not the full multiple sequence alignment.</p><p><code>opt</code> is currently ignored. It is kept for signature uniformity for other marshal functions</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L367-L375">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.marshal_json-Tuple{IO, PanGraph.Graphs.Graph}" href="#PanGraph.Graphs.marshal_json-Tuple{IO, PanGraph.Graphs.Graph}"><code>PanGraph.Graphs.marshal_json</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">marshal_json(io::IO, G::Graph; opt=nothing)</code></pre><p>Serialize graph <code>G</code> as a json format output stream <code>io</code>. This is the main storage/exported format for PanGraph. Currently it is the only format that can reconstruct an in-memory pangraph.</p><p><code>opt</code> is currently ignored. It is kept for signature uniformity for other marshal functions</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L383-L391">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.prune!-Tuple{PanGraph.Graphs.Graph}" href="#PanGraph.Graphs.prune!-Tuple{PanGraph.Graphs.Graph}"><code>PanGraph.Graphs.prune!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">prune!(G::Graph)</code></pre><p>Remove all blocks from graph <code>G</code> that are not currently used by any extant sequence. Internal function used during guide tree alignment.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L270-L275">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.purge!-Tuple{PanGraph.Graphs.Graph}" href="#PanGraph.Graphs.purge!-Tuple{PanGraph.Graphs.Graph}"><code>PanGraph.Graphs.purge!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">purge!(G::Graph)</code></pre><p>Remove all blocks from paths found in graph <code>G</code> that have zero length. Internal function used during guide tree alignment.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L281-L286">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.realign!-Tuple{PanGraph.Graphs.Graph}" href="#PanGraph.Graphs.realign!-Tuple{PanGraph.Graphs.Graph}"><code>PanGraph.Graphs.realign!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">realign!(G::Graph; accept)</code></pre><p>Realign blocks contained within graph <code>G</code>. Usage of this function requires <a href="https://mafft.cbrc.jp/alignment/software/source.html">MAFFT</a> to be on the system <strong>PATH</strong> <code>accept</code> should be a function that returns true on blocks you wish to realign. By default, all blocks are realigned.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L578-L585">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.sequence-Tuple{PanGraph.Graphs.Graph, AbstractString}" href="#PanGraph.Graphs.sequence-Tuple{PanGraph.Graphs.Graph, AbstractString}"><code>PanGraph.Graphs.sequence</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">sequence(G::Graph, name::String)</code></pre><p>Return the sequence corresponding to genome <code>name</code> within graph <code>G</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L559-L563">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.sequence-Tuple{PanGraph.Graphs.Graph}" href="#PanGraph.Graphs.sequence-Tuple{PanGraph.Graphs.Graph}"><code>PanGraph.Graphs.sequence</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">sequence(G::Graph)</code></pre><p>Return all pairs of <code>name</code> =&gt; <code>sequence</code> encoded within graph <code>G</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L571-L575">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.test" href="#PanGraph.Graphs.test"><code>PanGraph.Graphs.test</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">test(path)</code></pre><p>Align all sequences found in the fasta file at <code>path</code> into a pangraph. Verifies that after the alignment is complete, all sequences are correctly reconstructed</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L620-L625">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.unmarshal-Tuple{Any}" href="#PanGraph.Graphs.unmarshal-Tuple{Any}"><code>PanGraph.Graphs.unmarshal</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">unmarshal(io::IO)</code></pre><p>Deserialize the json formatted input stream <code>io</code> into a Graph data structure. Return a <code>Graph</code> type.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/graph.jl#L459-L464">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../edge/">« Edges</a><a class="docs-footer-nextpage" href="../gfa/">GFA »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.12 on <span class="colophon-date" title="Tuesday 15 February 2022 06:54">Tuesday 15 February 2022</span>. Using Julia version 1.7.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
