<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Polishing the pangraph and exploring alignments · PanGraph.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PanGraph.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorial</span><ul><li><a class="tocitem" href="../tutorial_1/">Building a pangraph</a></li><li><a class="tocitem" href="../tutorial_2/">The structure of Pangraph output file</a></li><li class="is-active"><a class="tocitem" href>Polishing the pangraph and exploring alignments</a><ul class="internal"><li><a class="tocitem" href="#Polishing-the-alignments"><span>Polishing the alignments</span></a></li><li><a class="tocitem" href="#Export-block-alignments-and-trees"><span>Export block alignments and trees</span></a></li><li><a class="tocitem" href="#Accessing-block-alignments-and-trees"><span>Accessing block alignments and trees</span></a></li><li><a class="tocitem" href="#Explore-block-alignments-with-the-panX-visualization"><span>Explore block alignments with the panX visualization</span></a></li></ul></li><li><a class="tocitem" href="../tutorial_4/">Projecting the graph on a subset of strains.</a></li><li><a class="tocitem" href="../tutorial_5/">Example application: plasmid rearrangements</a></li><li><a class="tocitem" href="../tutorial_6/">Example application: antibiotic resistance gene neighbourhoods</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/pangraph/">PanGraph globals</a></li><li><a class="tocitem" href="../../lib/align/">Alignment</a></li><li><a class="tocitem" href="../../lib/block/">Blocks</a></li><li><a class="tocitem" href="../../lib/edge/">Edges</a></li><li><a class="tocitem" href="../../lib/graph/">Graphs</a></li><li><a class="tocitem" href="../../lib/gfa/">GFA</a></li><li><a class="tocitem" href="../../lib/mash/">Mash Implementation</a></li><li><a class="tocitem" href="../../lib/minimap/">Minimap2 Wrapper</a></li><li><a class="tocitem" href="../../lib/mmseqs/">MMseqs Wrapper</a></li><li><a class="tocitem" href="../../lib/node/">Nodes</a></li><li><a class="tocitem" href="../../lib/path/">Paths</a></li><li><a class="tocitem" href="../../lib/simulate/">Simulation</a></li><li><a class="tocitem" href="../../lib/utility/">Utility</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/build/">Build</a></li><li><a class="tocitem" href="../../cli/export/">Export</a></li><li><a class="tocitem" href="../../cli/generate/">Generate</a></li><li><a class="tocitem" href="../../cli/marginalize/">Marginalize</a></li><li><a class="tocitem" href="../../cli/polish/">Polish</a></li><li><a class="tocitem" href="../../cli/version/">Version</a></li></ul></li><li><span class="tocitem">Development</span><ul><li><a class="tocitem" href="../../dev/building-docker/">Building Docker image</a></li><li><a class="tocitem" href="../../dev/building-documentation/">Updating the documentation</a></li><li><a class="tocitem" href="../../dev/releasing/">Releasing</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorial</a></li><li class="is-active"><a href>Polishing the pangraph and exploring alignments</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Polishing the pangraph and exploring alignments</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/neherlab/pangraph/blob/master/docs/src/tutorials/tutorial_3.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Polishing-the-pangraph-and-exploring-alignments"><a class="docs-heading-anchor" href="#Polishing-the-pangraph-and-exploring-alignments">Polishing the pangraph and exploring alignments</a><a id="Polishing-the-pangraph-and-exploring-alignments-1"></a><a class="docs-heading-anchor-permalink" href="#Polishing-the-pangraph-and-exploring-alignments" title="Permalink"></a></h1><p>This third tutorial is focused on block alignments and is divided in three main parts:</p><ul><li>refining the block alignments using the <code>polish</code> command.</li><li>exporting the alignment (and related phylogenetic tree) of each block.</li><li>exploring the full set of block alignments using the <a href="https://pangenome.org/">panX</a><sup class="footnote-reference"><a id="citeref-1" href="#footnote-1">[1]</a></sup> visualization.</li></ul><div class="admonition is-info"><header class="admonition-header">Requirements</header><div class="admonition-body"><p>Polishing the pangenome graph requires <a href="https://mafft.cbrc.jp/alignment/software/">mafft</a>. Exporting files for the panX visualization requires <a href="http://www.microbesonline.org/fasttree">fasttree</a>. Running the panX visualization requires <a href="https://nodejs.org/en/">node.js</a>.</p></div></div><h2 id="Polishing-the-alignments"><a class="docs-heading-anchor" href="#Polishing-the-alignments">Polishing the alignments</a><a id="Polishing-the-alignments-1"></a><a class="docs-heading-anchor-permalink" href="#Polishing-the-alignments" title="Permalink"></a></h2><p>The pangenome graph is built by iterative pairwise merges of smaller graphs along a guide tree. At each stage blocks are compared and aligned using only their <em>consensus</em>. This shortens considerably the time needed to build a pangraph, but might introduce minor inconsistencies and artifacts in the alignments. These can be removed using the <code>polish</code> command which reconstructs the sequences in the block and performs a multiple sequence alignment (e.g. using mafft) (see <a href="../../cli/polish/#Polish">Polish</a>).</p><p>To polish the <code>ecoli_pangraph.json</code> file that was produced in the first tutorial we can run the command:</p><pre><code class="language-bash hljs">pangraph polish ecoli_pangraph.json &gt; polished_pangraph.json</code></pre><p>This should complete in 10/15 minutes on 4 cores. Optionally, a threshold length can be specified with the flag <code>--length</code>. In this case only alignments of blocks whose consensus sequence is shorter than this threshold are refined.</p><p>Note that the only effect of this command is refining the alignments, but the &quot;topological&quot; structure of the pangraph is left unchanged. As such, polishing a pangraph is only useful when one is directly interested in block alignments.</p><h2 id="Export-block-alignments-and-trees"><a class="docs-heading-anchor" href="#Export-block-alignments-and-trees">Export block alignments and trees</a><a id="Export-block-alignments-and-trees-1"></a><a class="docs-heading-anchor-permalink" href="#Export-block-alignments-and-trees" title="Permalink"></a></h2><p>By running the <code>export</code> command with the <code>--export-panX</code> option (see <a href="../../cli/export/#Export">Export</a>) one can export all the block alignments, together with additional files that are needed for the panX visualization. For our example we can run:</p><pre><code class="language-bash hljs">pangraph export \
    --export-panX \
    --output-directory ecoli_export \
    --no-export-gfa \
    polished_pangraph.json</code></pre><p>The flag <code>--no-export-gfa</code> excludes the export in gfa format that we covered in the first tutorial. This command should run in about 15 minutes. It will generate a folder named <code>ecoli_export/vis</code> that contain multiple files. These include:</p><ul><li>the alignment of each block (<code>geneCluster/*_aln.fa.gz</code> files)</li><li>the phylogenetic tree generated from each alignment (<code>geneCluster/*.nwk</code> files)</li><li>the phylogenetic tree obtained from the concatenated alignment of all core blocks (<code>strain_tree.nwk</code>)</li><li>additional files needed for the panX visualization</li></ul><h2 id="Accessing-block-alignments-and-trees"><a class="docs-heading-anchor" href="#Accessing-block-alignments-and-trees">Accessing block alignments and trees</a><a id="Accessing-block-alignments-and-trees-1"></a><a class="docs-heading-anchor-permalink" href="#Accessing-block-alignments-and-trees" title="Permalink"></a></h2><p>The file <code>geneCluster.json</code> contains the index of all block alignments. It consists of a list with one entry per block. Here is an example of an entry:</p><pre><code class="language-json hljs">{
    &quot;GName&quot;: &quot;none&quot;,
    &quot;geneLen&quot;: 114,       // length of block in bp
    &quot;dup_detail&quot;: &quot;&quot;,
    &quot;geneId&quot;: 1129,
    &quot;allGName&quot;: &quot;none&quot;,
    &quot;count&quot;: 5,           // tot n. of block occurrences
    &quot;locus&quot;: [            // list of &quot;strain # node number&quot;
        &quot;NZ_CP010150#1&quot;,
        &quot;NZ_CP010183#1&quot;,
        &quot;NZ_CP015912#1&quot;,
        &quot;NC_013361#1&quot;,
        &quot;NZ_CP019944#1&quot;
    ],
    &quot;msa&quot;: &quot;RC00001129&quot;,  // Id assigned to the alignment.
    &quot;divers&quot;: 0.03508771929824561,
    &quot;ann&quot;: &quot;SXSJVKIOIY&quot;,  // block id
    &quot;dupli&quot;: &quot;no&quot;,        // whether the block is duplicated
    &quot;allAnn&quot;: &quot;SXSJVKIOIY&quot;
},</code></pre><p>The most important properties of this object are:<sup class="footnote-reference"><a id="citeref-2" href="#footnote-2">[2]</a></sup></p><ul><li><code>ann</code>: block id.</li><li><code>msa</code>: id assigned to the block alignment by the export function. This is the prefix of the corresponding alignment file.</li><li><code>geneLen</code>: length of the consensus sequence of the block.</li><li><code>count</code>: total number of occurrences of the block.</li><li><code>locus</code>: list of block occurrences (nodes). Each one is labeled as &quot;strain # node number&quot;.</li><li><code>dupli</code>: whether the block is duplicated in some strain.</li></ul><p>In particular, the <code>msa</code> property provides the prefix of the file containing the block alignment and the phylogenetic tree. In the example above, the nucleotide alignment for block <code>SXSJVKIOIY</code> can be found in file <code>geneCluster/RC00001129_na_aln.fa.gz</code>, and the corresponding phylogenetic tree is <code>RC00001129.nwk</code>.</p><h2 id="Explore-block-alignments-with-the-panX-visualization"><a class="docs-heading-anchor" href="#Explore-block-alignments-with-the-panX-visualization">Explore block alignments with the panX visualization</a><a id="Explore-block-alignments-with-the-panX-visualization-1"></a><a class="docs-heading-anchor-permalink" href="#Explore-block-alignments-with-the-panX-visualization" title="Permalink"></a></h2><p>This export format was designed to be compatible with the <a href="https://github.com/neherlab/pan-genome-visualization">panX visualization</a>. This visualization was originally created for exploring gene clusters, but it is equally informative if one uses block alignments instead of gene alignments. The panX repository linked above contains instructions on how to set up the visualization.</p><p>For our example we need to:</p><ul><li>clone the repository and perform the setup</li><li>add a new page to the visualization for our example data</li><li>copy the export file we previously created into the apposite folder</li><li>start the server</li></ul><p>this can be done with the command:</p><pre><code class="language-bash hljs"># clone the repository, access it and install packages
git clone https://github.com/neherlab/pan-genome-visualization
cd pan-genome-visualization
git submodule update --init
npm install

# add a new page for our dataset named Ecoli, with wide format
bash add-new-pages-repo.sh Ecoli wide

# copy pangraph exported files into the appropriate panx directory
cp -r ../ecoli_export/vis/* public/dataset/Ecoli/

# start the server
npm start</code></pre><p>Once the server has started, the visualization can be accessed at <a href="http://localhost:8000/Ecoli">http://localhost:8000/Ecoli</a>.</p><p><img src="../../assets/panx_visualization.png" alt="img"/></p><p>Here is a summary of the information displayed in different panels:</p><ul><li>A: rank plot of block counts. The initial peak is represented by duplicated blocks (more than 10 counts). Core blocks are found in the plateau at count equal to 10. Rare blocks populate the tail of the plot.</li><li>B: distribution of block lengths. The longest blocks in our example have length of more than 50 kbps.</li><li>C: table relating block id (<code>annotation</code>) to number of counts (<code>#Strains</code>), whether the block is duplicated, average diversity, and length of the consensus sequence in basepairs.</li><li>D: visualization of the block alignment for the block selected in panel C. Entries are labeled according to the strain and the number of block occurrence.</li><li>E: comparison of the core-block tree (left, phylogenetic tree constructed from the concatenated alignment of all core blocks) and the tree reconstructed from the alignment selected in panel C (right).</li></ul><p>This visualization makes it easy to quickly explore the data. For example if we order the blocks by number of counts we find that the most duplicated block has 86 counts and consensus length of 768 bp, From panel D we can easily download the alignment and if we run BLAST on the sequence we discover that it contains the genetic sequence of a transposase.</p><section class="footnotes is-size-7"><ul><li class="footnote" id="footnote-1"><a class="tag is-link" href="#citeref-1">1</a>Ding, Wei, Franz Baumdicker, and Richard A. Neher. &quot;panX: pan-genome analysis and exploration.&quot; Nucleic acids research 46.1 (2018): e5-e5.</li><li class="footnote" id="footnote-2"><a class="tag is-link" href="#citeref-2">2</a>The names of these entries was assigned so as to be compatible with the panX visualization, which uses genes instead of blocks. This is why for example the length is saved as <code>geneLen</code>, even though the pangenome graph and blocks do not contain gene annotations.</li></ul></section></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../tutorial_2/">« The structure of Pangraph output file</a><a class="docs-footer-nextpage" href="../tutorial_4/">Projecting the graph on a subset of strains. »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Monday 30 October 2023 09:00">Monday 30 October 2023</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
