<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Blocks · PanGraph.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><script src="../../../copy.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PanGraph.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorial</span><ul><li><a class="tocitem" href="../../tutorials/tutorial_1/">Building a pangraph</a></li><li><a class="tocitem" href="../../tutorials/tutorial_2/">The structure of Pangraph output file</a></li><li><a class="tocitem" href="../../tutorials/tutorial_3/">Polishing the pangraph and exploring alignments</a></li><li><a class="tocitem" href="../../tutorials/tutorial_4/">Projecting the graph on a subset of strains.</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../align/">Alignment</a></li><li class="is-active"><a class="tocitem" href>Blocks</a><ul class="internal"><li><a class="tocitem" href="#Types"><span>Types</span></a></li><li><a class="tocitem" href="#Functions"><span>Functions</span></a></li></ul></li><li><a class="tocitem" href="../edge/">Edges</a></li><li><a class="tocitem" href="../graph/">Graphs</a></li><li><a class="tocitem" href="../gfa/">GFA</a></li><li><a class="tocitem" href="../mash/">Mash Implementation</a></li><li><a class="tocitem" href="../minimap/">Minimap2 Wrapper</a></li><li><a class="tocitem" href="../node/">Nodes</a></li><li><a class="tocitem" href="../path/">Paths</a></li><li><a class="tocitem" href="../simulate/">Simulation</a></li><li><a class="tocitem" href="../utility/">Utility</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/build/">Build</a></li><li><a class="tocitem" href="../../cli/export/">Export</a></li><li><a class="tocitem" href="../../cli/generate/">Generate</a></li><li><a class="tocitem" href="../../cli/marginalize/">Marginalize</a></li><li><a class="tocitem" href="../../cli/polish/">Polish</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>Blocks</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Blocks</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/neherlab/pangraph/blob/master/docs/src/lib/block.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Blocks"><a class="docs-heading-anchor" href="#Blocks">Blocks</a><a id="Blocks-1"></a><a class="docs-heading-anchor-permalink" href="#Blocks" title="Permalink"></a></h1><h2 id="Types"><a class="docs-heading-anchor" href="#Types">Types</a><a id="Types-1"></a><a class="docs-heading-anchor-permalink" href="#Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.Block" href="#PanGraph.Graphs.Blocks.Block"><code>PanGraph.Graphs.Blocks.Block</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">mutable struct Block
    uuid     :: String
    sequence :: Array{UInt8}
    gaps     :: Dict{Int,Int}
    mutate   :: Dict{Node{Block},SNPMap}
    insert   :: Dict{Node{Block},InsMap}
    delete   :: Dict{Node{Block},DelMap}
end</code></pre><p>Store a multiple sequence alignment of contiguous DNA related by homology. Use as a component of a larger, branching multiple genome alignment. <code>uuid</code> is a string identifier unique to each block <code>sequence</code> is the consensus (majority-rule) sequence <code>gaps</code> recapitulate all locations of insertions for generating the full sequence alignment. <code>mutate</code>, <code>insert</code>, and <code>delete</code> store polymorphisms of each genome contained within the block.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L269-L285">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.Block-NTuple{5, Any}" href="#PanGraph.Graphs.Blocks.Block-NTuple{5, Any}"><code>PanGraph.Graphs.Blocks.Block</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Block(sequence,gaps,mutate,insert,delete)</code></pre><p>Construct a block with a unique <code>uuid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L309-L313">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.Block-Tuple{Any, Any}" href="#PanGraph.Graphs.Blocks.Block-Tuple{Any, Any}"><code>PanGraph.Graphs.Blocks.Block</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Block(sequence,gaps)</code></pre><p>Construct a block with a unique <code>uuid</code> with fixed <code>sequence</code> and <code>gaps</code>. No individuals and thus polymorphisms are initialized.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L315-L320">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.Block-Tuple{Any}" href="#PanGraph.Graphs.Blocks.Block-Tuple{Any}"><code>PanGraph.Graphs.Blocks.Block</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Block(sequence,gaps)</code></pre><p>Construct a block with a unique <code>uuid</code> with fixed <code>sequence</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L322-L326">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.Block-Tuple{PanGraph.Graphs.Blocks.Block, Any}" href="#PanGraph.Graphs.Blocks.Block-Tuple{PanGraph.Graphs.Blocks.Block, Any}"><code>PanGraph.Graphs.Blocks.Block</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Block(b::Block, slice)</code></pre><p>Return a subsequence associated to block <code>b</code> at interval <code>slice</code>. The returned block has a newly generated <code>uuid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L474-L479">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.Block-Tuple{Vararg{PanGraph.Graphs.Blocks.Block}}" href="#PanGraph.Graphs.Blocks.Block-Tuple{Vararg{PanGraph.Graphs.Blocks.Block}}"><code>PanGraph.Graphs.Blocks.Block</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Block(bs::Block...)</code></pre><p>Concatenate a variable number of blocks into one larger block. The returned block has a newly generated <code>uuid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L441-L446">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.Block-Tuple{}" href="#PanGraph.Graphs.Blocks.Block-Tuple{}"><code>PanGraph.Graphs.Blocks.Block</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Block(sequence,gaps)</code></pre><p>Construct a block with a unique <code>uuid</code>. All fields are empty.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L328-L332">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.PairPos" href="#PanGraph.Graphs.Blocks.PairPos"><code>PanGraph.Graphs.Blocks.PairPos</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">mutable struct PairPos
    qry :: Maybe{Pos}
    ref :: Maybe{Pos}
end</code></pre><p>Representation of matched pair of intervals within a pairwise alignment. <code>qry</code> can be of type <code>Pos</code> or <code>Nothing</code> <code>ref</code> can be of type <code>Pos</code> or <code>Nothing</code> If either <code>ref</code> or <code>qry</code> is nothing, the PairPos corresponds to an insertion or deletion respectively.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L112-L122">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.Pos" href="#PanGraph.Graphs.Blocks.Pos"><code>PanGraph.Graphs.Blocks.Pos</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">mutable struct Pos
    start :: Int
    stop  :: Int
end</code></pre><p>Representation of a single interval within a pairwise alignment. Inclusive on both ends, i.e. includes <code>start</code> and <code>stop</code> Used internally to unpack cigar strings.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L93-L102">source</a></section></article><h2 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Base.append!-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}, Union{Nothing, Dict{Int64, UInt8}}, Union{Nothing, Dict{Tuple{Int64, Int64}, Vector{UInt8}}}, Union{Nothing, Dict{Int64, Int64}}}" href="#Base.append!-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}, Union{Nothing, Dict{Int64, UInt8}}, Union{Nothing, Dict{Tuple{Int64, Int64}, Vector{UInt8}}}, Union{Nothing, Dict{Int64, Int64}}}"><code>Base.append!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">append!(b::Block, node::Node{Block}, snp::Maybe{SNPMap}, ins::Maybe{InsMap}, del::Maybe{DelMap})</code></pre><p>Adds a new genome at <code>node</code> to multiple sequence alignment block <code>b</code>. Polymorphisms are optional. If nothing is passed instead, an empty dictionary will be used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L808-L813">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.length-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node}" href="#Base.length-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node}"><code>Base.length</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">length(b::Block, n::Node)</code></pre><p>Return the length of the sequence of node <code>n</code> within the multiple alignment of block <code>b</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L519-L523">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.length-Tuple{PanGraph.Graphs.Blocks.Block}" href="#Base.length-Tuple{PanGraph.Graphs.Blocks.Block}"><code>Base.length</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">length(b::Block)</code></pre><p>Return the length of consensus sequence of the multiple alignment of block <code>b</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L513-L517">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.pop!-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node}" href="#Base.pop!-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node}"><code>Base.pop!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">pop!(b::Block, n::Node)</code></pre><p>Remove genome of Node <code>n</code> from Block <code>b</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L1402-L1406">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.allele_positions-Tuple{Dict{Int64, UInt8}, Dict{Tuple{Int64, Int64}, Vector{UInt8}}, Dict{Int64, Int64}}" href="#PanGraph.Graphs.Blocks.allele_positions-Tuple{Dict{Int64, UInt8}, Dict{Tuple{Int64, Int64}, Vector{UInt8}}, Dict{Int64, Int64}}"><code>PanGraph.Graphs.Blocks.allele_positions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">allele_positions(snp::SNPMap, ins::InsMap, del::DelMap)</code></pre><p>Return an iterator over polymorphic loci, i.e. SNPs and Indels. The iterator will be sorted by position in ascending order.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L556-L561">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.allele_positions-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node}" href="#PanGraph.Graphs.Blocks.allele_positions-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node}"><code>PanGraph.Graphs.Blocks.allele_positions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">allele_positions(b::Block, n::Node)</code></pre><p>Return an iterator over polymorphic loci for node <code>n</code> contained within block <code>b</code> The iterator will be sorted by position in ascending order.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L569-L574">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.applyalleles-NTuple{4, Any}" href="#PanGraph.Graphs.Blocks.applyalleles-NTuple{4, Any}"><code>PanGraph.Graphs.Blocks.applyalleles</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">applyalleles(seq::Array{UInt8}, mutate::SNPMap, insert::InsMap, delete::DelMap)</code></pre><p>Take a sequence and apply polymorphisms, as given by <code>mutate</code>, <code>insert</code>, and <code>delete</code>. Return the brand new allocated sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L36-L41">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.assert_equals-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Blocks.Block}" href="#PanGraph.Graphs.Blocks.assert_equals-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Blocks.Block}"><code>PanGraph.Graphs.Blocks.assert_equals</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">assert_equal(b₁::Block, b₂::Block)</code></pre><p>Throw an error in block <code>b₁</code> is not equivalent to block <code>b₂</code>. Useful for internal debugging.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L600-L605">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.check-Tuple{PanGraph.Graphs.Blocks.Block}" href="#PanGraph.Graphs.Blocks.check-Tuple{PanGraph.Graphs.Blocks.Block}"><code>PanGraph.Graphs.Blocks.check</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">check(b::Block)</code></pre><p>Check whether block <code>b</code> is internally self-consistent. Useful for debugging internals.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L1294-L1299">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.combine-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Utility.Alignment}" href="#PanGraph.Graphs.Blocks.combine-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Utility.Alignment}"><code>PanGraph.Graphs.Blocks.combine</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">combine(qry::Block, ref::Block, aln::Alignment; minblock=500)</code></pre><p>Take a pairwise alignment <code>aln</code> from the consensus of <code>qry</code> to <code>ref</code> and merge both. The resultant new block, with a novel uuid is returned. Alignment <code>aln</code> is a segmented set of intervals mapping homologous regions of one block into the other. Parameter <code>minblock</code> is the cutoff length of an indel, above which a new block will be created.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L1252-L1259">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.depth-Tuple{PanGraph.Graphs.Blocks.Block}" href="#PanGraph.Graphs.Blocks.depth-Tuple{PanGraph.Graphs.Blocks.Block}"><code>PanGraph.Graphs.Blocks.depth</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">depth(b::Block)</code></pre><p>Return the number of genomes contained within the alignment</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L503-L507">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.diversity-Tuple{PanGraph.Graphs.Blocks.Block}" href="#PanGraph.Graphs.Blocks.diversity-Tuple{PanGraph.Graphs.Blocks.Block}"><code>PanGraph.Graphs.Blocks.diversity</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">diversity(b::Block)</code></pre><p>Return the averaged fraction of loci that are mutated within the multiple sequence alignment of block <code>b</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L530-L534">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.partition-Tuple{Any}" href="#PanGraph.Graphs.Blocks.partition-Tuple{Any}"><code>PanGraph.Graphs.Blocks.partition</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">partition(alignment; minblock=500)</code></pre><p>Parse the alignment into matched intervals of a pairwise alignment. If any insertion or deletion is larger than <code>minblock</code>, a new block is created to hold the homologous interval. This ensures that all blocks are at least <code>minblock</code> long <strong>and</strong> no block contains an insertion or deletion longer than itself.</p><p><code>alignment</code> is assumed to be an data structure from the Utility module</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L131-L139">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.reconsensus!-Tuple{PanGraph.Graphs.Blocks.Block}" href="#PanGraph.Graphs.Blocks.reconsensus!-Tuple{PanGraph.Graphs.Blocks.Block}"><code>PanGraph.Graphs.Blocks.reconsensus!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">reconsensus!(b::Block)</code></pre><p>Update the consensus sequence of block <code>b</code> by majority-rule over the multiple sequence alignment.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L931-L935">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.regap!-Tuple{PanGraph.Graphs.Blocks.Block}" href="#PanGraph.Graphs.Blocks.regap!-Tuple{PanGraph.Graphs.Blocks.Block}"><code>PanGraph.Graphs.Blocks.regap!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">regap!(b::Block)</code></pre><p>Recompute the positions of gaps within the multiple sequence of block <code>b</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L900-L904">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.rereference-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Blocks.Block, Any}" href="#PanGraph.Graphs.Blocks.rereference-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Blocks.Block, Any}"><code>PanGraph.Graphs.Blocks.rereference</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">rerefence(qry::Block, ref::Block, aligment)</code></pre><p>Take a pairwise alignment <code>segments</code> from the consensus of <code>qry</code> to <code>ref</code> and rereference all polymorphisms of <code>qry</code> to the consensus sequence of `ref. Low-level function used by higher-level API.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L957-L963">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.swap!-Tuple{PanGraph.Graphs.Blocks.Block, Array{PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}}, PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}}" href="#PanGraph.Graphs.Blocks.swap!-Tuple{PanGraph.Graphs.Blocks.Block, Array{PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}}, PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}}"><code>PanGraph.Graphs.Blocks.swap!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">swap!(b::Block, oldkey::Array{Node{Block}}, newkey::Node{Block})</code></pre><p>Remove all polymorphisms associated to all keys within <code>oldkey</code>. Concatenate and reassociate them to <code>newkey</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L845-L850">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Blocks.swap!-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}, PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}}" href="#PanGraph.Graphs.Blocks.swap!-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}, PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}}"><code>PanGraph.Graphs.Blocks.swap!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">swap!(b::Block, oldkey::Node{Block}, newkey::Node{Block})</code></pre><p>Remove all polymorphisms associated to <code>oldkey</code> and reassociate them to <code>newkey</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L834-L838">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.marshal_fasta-Tuple{IO, PanGraph.Graphs.Blocks.Block}" href="#PanGraph.Graphs.marshal_fasta-Tuple{IO, PanGraph.Graphs.Blocks.Block}"><code>PanGraph.Graphs.marshal_fasta</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">marshal_fasta(io::IO, b::Block; opt=nothing)</code></pre><p>Serialize the multiple sequence alignment of block <code>b</code> to a fasta format to IO stream <code>io</code>. Each sequence will be serialized as-is, i.e. with no gaps.</p><p>If <code>opt</code> is not <code>nothing</code>, the output will be an aligned fasta file. Futhermore, opt is interpreted as a function to be called per internal <code>node</code> that gives a unique name for each fasta record that is generated per <code>node</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L1374-L1383">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.reverse_complement-Tuple{PanGraph.Graphs.Blocks.Block}" href="#PanGraph.Graphs.reverse_complement-Tuple{PanGraph.Graphs.Blocks.Block}"><code>PanGraph.Graphs.reverse_complement</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">reverse_complement(b::Block; keepid=false)</code></pre><p>Return the reverse complement of the multiple sequence alignment within Block <code>b</code>. By default, will return a block with a new <code>uuid</code>, unless keepid is set to <code>true</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L578-L583">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.sequence!-Tuple{Any, PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}}" href="#PanGraph.Graphs.sequence!-Tuple{Any, PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}}"><code>PanGraph.Graphs.sequence!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">sequence!(seq, b::Block, node::Node{Block}; gaps=false)</code></pre><p>Mutate the sequence buffer <code>seq</code> in place to hold the sequence associated to genome <code>node</code> within sequence alignment of block <code>b</code>. By default, gaps (charater &#39;-&#39;) will not be returned, unless <code>gaps</code> is set to <code>true</code>. Return the sequence with gap characters to generate the full sequence alignment.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L693-L699">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.sequence-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}}" href="#PanGraph.Graphs.sequence-Tuple{PanGraph.Graphs.Blocks.Block, PanGraph.Graphs.Nodes.Node{PanGraph.Graphs.Blocks.Block}}"><code>PanGraph.Graphs.sequence</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">sequence(seq, b::Block, node::Node{Block}; gaps=false, forward=false)</code></pre><p>Return the sequence associated to genome <code>node</code> within sequence alignment of block <code>b</code>. By default, gaps (charater &#39;-&#39;) will not be returned, unless <code>gaps</code> is set to <code>true</code>. Return the sequence with gap characters can be used to generate the full sequence alignment. If <code>forward</code> is true, the true orientation of the genome is ignored and will be returned to align to the forward consensus.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L766-L773">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.sequence-Tuple{PanGraph.Graphs.Blocks.Block}" href="#PanGraph.Graphs.sequence-Tuple{PanGraph.Graphs.Blocks.Block}"><code>PanGraph.Graphs.sequence</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">sequence(b::Block; gaps=false)</code></pre><p>Return the consensus of the multiple sequence alignment within block <code>b</code>. By default, gaps (charater &#39;-&#39;) will not be returned, unless <code>gaps</code> is set to <code>true</code>. Return the consensus alignment with gaps is useful for generating the full sequence alignment.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/0ca93b323cb5b67fb3097d69c2f4f38f60ede21f/src/block.jl#L614-L620">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../align/">« Alignment</a><a class="docs-footer-nextpage" href="../edge/">Edges »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.12 on <span class="colophon-date" title="Tuesday 15 February 2022 06:54">Tuesday 15 February 2022</span>. Using Julia version 1.7.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
