<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Example application: antibiotic resistance gene neighbourhoods · PanGraph.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PanGraph.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorial</span><ul><li><a class="tocitem" href="../tutorial_1/">Building a pangraph</a></li><li><a class="tocitem" href="../tutorial_2/">The structure of Pangraph output file</a></li><li><a class="tocitem" href="../tutorial_3/">Polishing the pangraph and exploring alignments</a></li><li><a class="tocitem" href="../tutorial_4/">Projecting the graph on a subset of strains.</a></li><li><a class="tocitem" href="../tutorial_5/">Example application: plasmid rearrangements</a></li><li class="is-active"><a class="tocitem" href>Example application: antibiotic resistance gene neighbourhoods</a><ul class="internal"><li><a class="tocitem" href="#Dataset"><span>Dataset</span></a></li><li><a class="tocitem" href="#Running-pangraph"><span>Running pangraph</span></a></li><li><a class="tocitem" href="#Linear-visualization"><span>Linear visualization</span></a></li></ul></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/pangraph/">PanGraph globals</a></li><li><a class="tocitem" href="../../lib/align/">Alignment</a></li><li><a class="tocitem" href="../../lib/block/">Blocks</a></li><li><a class="tocitem" href="../../lib/edge/">Edges</a></li><li><a class="tocitem" href="../../lib/graph/">Graphs</a></li><li><a class="tocitem" href="../../lib/gfa/">GFA</a></li><li><a class="tocitem" href="../../lib/mash/">Mash Implementation</a></li><li><a class="tocitem" href="../../lib/minimap/">Minimap2 Wrapper</a></li><li><a class="tocitem" href="../../lib/mmseqs/">MMseqs Wrapper</a></li><li><a class="tocitem" href="../../lib/node/">Nodes</a></li><li><a class="tocitem" href="../../lib/path/">Paths</a></li><li><a class="tocitem" href="../../lib/simulate/">Simulation</a></li><li><a class="tocitem" href="../../lib/utility/">Utility</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/build/">Build</a></li><li><a class="tocitem" href="../../cli/export/">Export</a></li><li><a class="tocitem" href="../../cli/generate/">Generate</a></li><li><a class="tocitem" href="../../cli/marginalize/">Marginalize</a></li><li><a class="tocitem" href="../../cli/polish/">Polish</a></li></ul></li><li><span class="tocitem">Development</span><ul><li><a class="tocitem" href="../../dev/building-docker/">Building Docker image</a></li><li><a class="tocitem" href="../../dev/building-documentation/">Updating the documentation</a></li><li><a class="tocitem" href="../../dev/releasing/">Releasing</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorial</a></li><li class="is-active"><a href>Example application: antibiotic resistance gene neighbourhoods</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Example application: antibiotic resistance gene neighbourhoods</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/neherlab/pangraph/blob/master/docs/src/tutorials/tutorial_6.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Example-application:-antibiotic-resistance-gene-neighbourhoods"><a class="docs-heading-anchor" href="#Example-application:-antibiotic-resistance-gene-neighbourhoods">Example application: antibiotic resistance gene neighbourhoods</a><a id="Example-application:-antibiotic-resistance-gene-neighbourhoods-1"></a><a class="docs-heading-anchor-permalink" href="#Example-application:-antibiotic-resistance-gene-neighbourhoods" title="Permalink"></a></h1><p>Regions around antibiotic resistance genes are often structurally diverse, due to the high numbers of rearrangements e.g. those arising from repeated insertions of different insertion sequences (ISs). In this tutorial we&#39;ll use pangraph to explore a dataset of isolates which all contain a beta-lactamase gene: KPC-2 or KPC-3 (1 SNP difference). Rather than look at the whole genomes, we will explore just the immediate neighbourhood of the KPC gene.</p><h2 id="Dataset"><a class="docs-heading-anchor" href="#Dataset">Dataset</a><a id="Dataset-1"></a><a class="docs-heading-anchor-permalink" href="#Dataset" title="Permalink"></a></h2><p>We will be using a set of n=34 isolates from David et al. (2020) (<a href="https://pubmed.ncbi.nlm.nih.gov/32968015/">paper</a>). This is a small part of a larger dataset of KPC-carrying <em>Klebsiella pneumoniae</em> isolated from patients across European hospitals.</p><p>Theoretically you could run pangraph on the full genomes of the isolates. However, in this case we are only interested in the region immediately around KPC. So to prepare the data for this tutorial we have found the gene (with <code>blast</code>) and then extracted the surrounding region: 10kb upstream and 5kb downstream.<sup class="footnote-reference"><a id="citeref-1" href="#footnote-1">[1]</a></sup></p><p>You can download this dataset by running the following:</p><pre><code class="language-bash hljs">wget https://raw.githubusercontent.com/liampshaw/pangraph-tutorials/main/data/kpc/kpc-contigs-u10k-d5k.fa</code></pre><h2 id="Running-pangraph"><a class="docs-heading-anchor" href="#Running-pangraph">Running pangraph</a><a id="Running-pangraph-1"></a><a class="docs-heading-anchor-permalink" href="#Running-pangraph" title="Permalink"></a></h2><p>We can now run pangraph on these extracted regions (n=34). This should take only a few seconds to run.</p><pre><code class="language-bash hljs">pangraph build kpc-contigs-u10k-d5k.fa &gt; pangraph_kpc_u10k_d5k.json
pangraph export --edge-minimum-length 0 pangraph_kpc_u10k_d5k.json -p pangraph_kpc_u10k_d5k -o ./</code></pre><p>Pangraph gives us three forms of output:</p><ul><li><code>pangraph_kpc_u10k_d5k.json</code> - json file storing the whole graph</li><li><code>pangraph_kpc_u10k_d5k.gfa</code> - graph in <a href="http://gfa-spec.github.io/GFA-spec/GFA1.html">GFA</a> format</li><li><code>pangraph_kpc_u10k_d5k.fa</code> - multifasta containing the consensus sequences of the pangenome blocks</li></ul><p>We know by construction that the KPC gene should be in all the contigs, so should be in the same alignment block in all sequences. If we download the KPC gene, we can then use <code>blast</code> to find this block from the fasta file with the consensus sequences of the blocks.</p><pre><code class="language-bash hljs">wget https://raw.githubusercontent.com/liampshaw/pangraph-tutorials/main/data/kpc/kpc2.fa
makeblastdb -in pangraph_kpc_u10k_d5k.fa -dbtype &#39;nucl&#39;
geneBlock=$(blastn -query kpc2.fa -db pangraph_kpc_u10k_d5k.fa -outfmt 6 | cut -f 2)
echo $geneBlock</code></pre><p>The gene will be on a block with an ID that is a random string (this string will be different if you rerun pangraph; the actual block will be the same).</p><p>Similar to the previous tutorial, we then convert the gfa into a csv that stores each contig as a linear path of blocks with their start and end positions.<sup class="footnote-reference"><a id="citeref-2" href="#footnote-2">[2]</a></sup></p><pre><code class="language-bash hljs"># Download custom script
wget https://raw.githubusercontent.com/liampshaw/pangraph-tutorials/main/scripts/prepare-pangraph-gfa.py
python prepare-pangraph-gfa.py pangraph_kpc_u10k_d5k.gfa</code></pre><p>This makes three output files:</p><ul><li><code>${input}.blocks.csv</code> - dataset of genome and block start/end positions</li><li><code>${input}.colours.csv</code> - blocks with colours (hex codes). By default, blocks that only appear once in the dataset are coloured grey, others get random colours (passing <code>--all</code> to the script would colour all blocks)</li><li><code>${input}.coloured.gfa</code> - a gfa with the block colours added as an extra field</li></ul><p>We can then plot a graph with Bandage. Using the <code>--colour custom</code> flag will use the block colours we have added.</p><pre><code class="nohighlight hljs">Bandage image pangraph_kpc_u10k_d5k.gfa.coloured.gfa \
	pangraph_kpc_u10k_d5k.gfa.png \
	--height 4000 --width 7000 \
	--colour custom</code></pre><p>You should get something like this:</p><p><img src="../../assets/kpc_graph.png" alt="img"/></p><p>Here each block has a random colour. If a block only appears in a single contig, it is grey. We have added an arrow showing the block which contains KPC.<sup class="footnote-reference"><a id="citeref-3" href="#footnote-3">[3]</a></sup></p><p>From this visualization, we can see that most of these KPC-positive have a very similar structure in the flanking regions. The splitting of the graph into &#39;frayed ends&#39; at the edges of the image shows that this structure has been inserted into different genomic backgrounds. This pattern is characteristic of a transposon. Indeed, KPC is mobilized by the transposon <a href="https://pubmed.ncbi.nlm.nih.gov/21844325/">Tn4401</a>.</p><h2 id="Linear-visualization"><a class="docs-heading-anchor" href="#Linear-visualization">Linear visualization</a><a id="Linear-visualization-1"></a><a class="docs-heading-anchor-permalink" href="#Linear-visualization" title="Permalink"></a></h2><p>The graph visualization with Bandage is helpful, but it can also be useful to view the unique paths through the graph as a linear visualization. In this section we will use an R script to do this, as in the previous tutorial.  </p><pre><code class="language-bash hljs">wget https://raw.githubusercontent.com/liampshaw/pangraph-tutorials/main/scripts/plot-blocks.R
Rscript plot-blocks.R \
	pangraph_kpc_u10k_d5k.gfa.blocks.csv \
	$geneBlock pangraph_kpc_u10k_d5k.gfa.png \
	pangraph_kpc_plot.pdf</code></pre><p><img src="../../assets/linear_and_graph_kpc.png" alt="img"/></p><p>If you pick a genome on the left of the plot, you should be able to follow its path through the graph representation on the right using the colours.<sup class="footnote-reference"><a id="citeref-4" href="#footnote-4">[4]</a></sup> The block starting at position 0 is the KPC-block.</p><section class="footnotes is-size-7"><ul><li class="footnote" id="footnote-1"><a class="tag is-link" href="#citeref-1">1</a>David et al. actually long-read sequenced n=44 isolates. For simplicity, we have only kept contigs that were sufficiently long, hence 34.</li><li class="footnote" id="footnote-2"><a class="tag is-link" href="#citeref-2">2</a>We are using the consensus lengths of the block for this example. It would be better to use the actual lengths of the block in each contig.</li><li class="footnote" id="footnote-3"><a class="tag is-link" href="#citeref-3">3</a>If you want to see the names of the nodes to identify the KPC-containing block in your visualization, you can add <code>--names --toutline 3 --fontsize 16</code> to the above Bandage command.</li><li class="footnote" id="footnote-4"><a class="tag is-link" href="#citeref-4">4</a>You may find that your graph plot and linear plot have left and right flipped. This is due to the random layout chosen by Bandage.</li></ul></section></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../tutorial_5/">« Example application: plasmid rearrangements</a><a class="docs-footer-nextpage" href="../../lib/pangraph/">PanGraph globals »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Thursday 13 October 2022 03:24">Thursday 13 October 2022</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
