<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Example application: plasmid rearrangements · PanGraph.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PanGraph.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorial</span><ul><li><a class="tocitem" href="../tutorial_1/">Building a pangraph</a></li><li><a class="tocitem" href="../tutorial_2/">The structure of Pangraph output file</a></li><li><a class="tocitem" href="../tutorial_3/">Polishing the pangraph and exploring alignments</a></li><li><a class="tocitem" href="../tutorial_4/">Projecting the graph on a subset of strains.</a></li><li class="is-active"><a class="tocitem" href>Example application: plasmid rearrangements</a><ul class="internal"><li><a class="tocitem" href="#Preliminary-steps"><span>Preliminary steps</span></a></li><li><a class="tocitem" href="#Default-visualization"><span>Default visualization</span></a></li><li><a class="tocitem" href="#Improving-the-visualization"><span>Improving the visualization</span></a></li></ul></li><li><a class="tocitem" href="../tutorial_6/">Example application: antibiotic resistance gene neighbourhoods</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/pangraph/">PanGraph globals</a></li><li><a class="tocitem" href="../../lib/align/">Alignment</a></li><li><a class="tocitem" href="../../lib/block/">Blocks</a></li><li><a class="tocitem" href="../../lib/edge/">Edges</a></li><li><a class="tocitem" href="../../lib/graph/">Graphs</a></li><li><a class="tocitem" href="../../lib/gfa/">GFA</a></li><li><a class="tocitem" href="../../lib/mash/">Mash Implementation</a></li><li><a class="tocitem" href="../../lib/minimap/">Minimap2 Wrapper</a></li><li><a class="tocitem" href="../../lib/mmseqs/">MMseqs Wrapper</a></li><li><a class="tocitem" href="../../lib/node/">Nodes</a></li><li><a class="tocitem" href="../../lib/path/">Paths</a></li><li><a class="tocitem" href="../../lib/simulate/">Simulation</a></li><li><a class="tocitem" href="../../lib/utility/">Utility</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/build/">Build</a></li><li><a class="tocitem" href="../../cli/export/">Export</a></li><li><a class="tocitem" href="../../cli/generate/">Generate</a></li><li><a class="tocitem" href="../../cli/marginalize/">Marginalize</a></li><li><a class="tocitem" href="../../cli/polish/">Polish</a></li><li><a class="tocitem" href="../../cli/version/">Version</a></li></ul></li><li><span class="tocitem">Development</span><ul><li><a class="tocitem" href="../../dev/building-docker/">Building Docker image</a></li><li><a class="tocitem" href="../../dev/building-documentation/">Updating the documentation</a></li><li><a class="tocitem" href="../../dev/releasing/">Releasing</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorial</a></li><li class="is-active"><a href>Example application: plasmid rearrangements</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Example application: plasmid rearrangements</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/neherlab/pangraph/blob/master/docs/src/tutorials/tutorial_5.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Example-application:-plasmid-rearrangements"><a class="docs-heading-anchor" href="#Example-application:-plasmid-rearrangements">Example application: plasmid rearrangements</a><a id="Example-application:-plasmid-rearrangements-1"></a><a class="docs-heading-anchor-permalink" href="#Example-application:-plasmid-rearrangements" title="Permalink"></a></h1><p>These next two tutorials cover other possible applications of pangraph and processing the downstream output.</p><p>Although pangraph was developed with whole genomes in mind, it can be applied to other situations. In this tutorial, we will explore the structural diversity in some closely-related plasmids from a hospital outbreak. Pangraph can be applied to these &#39;as if&#39; they were whole genomes.  </p><h2 id="Preliminary-steps"><a class="docs-heading-anchor" href="#Preliminary-steps">Preliminary steps</a><a id="Preliminary-steps-1"></a><a class="docs-heading-anchor-permalink" href="#Preliminary-steps" title="Permalink"></a></h2><p>This tutorial uses a dataset of five closely-related plasmids. They were analysed previously by Sheppard et al. (2016) in a <a href="https://doi.org/10.1128/AAC.00464-16">paper</a> studying an outbreak of carbapenem-resistant bacteria in a hospital in Virginia, USA. These plasmids are all similar to an <a href="https://www.ncbi.nlm.nih.gov/nuccore/CP017937.1">index plasmid</a> from the hospital, but have some structural changes. We will show how pangraph output can be used to visualize this structural diversity.</p><p>You can download these sequences by running:</p><pre><code class="language-bash hljs">wget https://github.com/liampshaw/pangraph-tutorials/raw/main/data/sheppard/UVA01_plasmids.fa.gz</code></pre><p>Building the pangraph and exporting it for visualization is done with these commands (should be very quick as we are using plasmids, which are much smaller than whole genomes):</p><pre><code class="language-bash hljs">pangraph build --circular UVA01_plasmids.fa.gz &gt; UVA01_plasmids_pangraph.json
pangraph export --edge-minimum-length 0 UVA01_plasmids_pangraph.json -p UVA01_plasmids_pangraph -o ./</code></pre><p>We use <code>--edge-minimum-length 0</code> because we want to see all blocks.</p><h2 id="Default-visualization"><a class="docs-heading-anchor" href="#Default-visualization">Default visualization</a><a id="Default-visualization-1"></a><a class="docs-heading-anchor-permalink" href="#Default-visualization" title="Permalink"></a></h2><p>As before, we can visualize the output file <code>UVA01_plasmids_pangraph.gfa</code> with <a href="https://rrwick.github.io/Bandage/">Bandage</a>.</p><p><img src="../../assets/plasmid_example_pangraph.png" alt="img"/></p><p>Here, the node colour represents the depth of the blocks. However, it is difficult from this visualization to understand individual paths through the graph.</p><h2 id="Improving-the-visualization"><a class="docs-heading-anchor" href="#Improving-the-visualization">Improving the visualization</a><a id="Improving-the-visualization-1"></a><a class="docs-heading-anchor-permalink" href="#Improving-the-visualization" title="Permalink"></a></h2><p>We can use some custom scripts to look at representations of the plasmids alongside their pangraph. These scripts are not part of pangraph but are an example of how to process the output into visualizations. You can download them by running:</p><pre><code class="language-bash hljs">wget https://raw.githubusercontent.com/liampshaw/pangraph-tutorials/main/scripts/prepare-pangraph-gfa.py
wget https://raw.githubusercontent.com/liampshaw/pangraph-tutorials/main/scripts/plot-blocks-UVA01.R</code></pre><p>First, we run a script to generate random colours for the blocks.</p><pre><code class="language-bash hljs">python prepare-pangraph-gfa.py UVA01_plasmids.pangraph.gfa --all</code></pre><p>This produces three files:</p><ul><li><code>${input}.blocks.csv</code> - dataset of genome and block start/end positions, with block colours (hex codes)</li><li><code>${input}.colours.csv</code> - blocks with colours (hex codes). The <code>--all</code> flag in the python script means we give all blocks a colour, even those which only appear once.</li><li><code>${input}.coloured.gfa</code> - a gfa with the same block colours added as an extra field</li></ul><p>Then we call Bandage programmaticaly with the <code>custom</code> colours option on the coloured gfa:</p><pre><code class="language-bash hljs">Bandage image UVA01_plasmids_pangraph.gfa.coloured.gfa pangraph.png --height 2000 --width 2500 --colour custom --depth --lengths --fontsize 30 --nodewidth 8</code></pre><p>We can now look at a linear representation of the plasmids alongside this graph. The details are in the script, but we basically represent the plasmids linearly and choose a random block to &#39;anchor&#39; them all to the same place.  </p><pre><code class="language-bash hljs">Rscript plot-blocks-UVA01.R UVA01_plasmids_pangraph.gfa.blocks.csv pangraph.png pangraph_linear_vs_graph.pdf</code></pre><p>This produces a version<sup class="footnote-reference"><a id="citeref-1" href="#footnote-1">[1]</a></sup> where we can see the different &#39;walks&#39; taken through the graph by each plasmid. For ease of understanding, on the left-hand side we have added simplified labels of the blocks (a-f) since some of the blocks are very small and hard to make out otherwise.</p><p><img src="../../assets/plasmid_example_pangraph_linear.png" alt="img"/></p><p>Note that the plasmids are circular: in the linear representation, in two cases block &#39;a&#39; appears on the left rather than on the right, but the structure is identical.</p><p>Sheppard et al. previously analysed these plasmids and made a table of the structural changes they identified with respect to the reference plasmid (Table 1 in their <a href="https://doi.org/10.1128/AAC.00464-16">paper</a>). We can see that the structural changes they identified are interpretable from the visualization we have made:</p><table><tr><th style="text-align: right">Accession</th><th style="text-align: right">Species</th><th style="text-align: right">Date</th><th style="text-align: right">Length (bp)</th><th style="text-align: right">Structural differences (Sheppard)</th><th style="text-align: right">Corresponding block(s) (pangraph)</th></tr><tr><td style="text-align: right">NZ_CP011575.1</td><td style="text-align: right"><em>K. pneumoniae</em></td><td style="text-align: right">Mar-2011</td><td style="text-align: right">43,621</td><td style="text-align: right"></td><td style="text-align: right"></td></tr><tr><td style="text-align: right">NZ_CP011582.1</td><td style="text-align: right"><em>E. cloacae</em></td><td style="text-align: right">Aug-2012</td><td style="text-align: right">43,433</td><td style="text-align: right">188-bp deletion</td><td style="text-align: right"><code>f</code></td></tr><tr><td style="text-align: right">NZ_CP011598.1</td><td style="text-align: right"><em>K. intermedia</em></td><td style="text-align: right">Sep-2009</td><td style="text-align: right">43,621</td><td style="text-align: right"></td><td style="text-align: right"></td></tr><tr><td style="text-align: right">NZ_CP011608.1</td><td style="text-align: right"><em>C. freundii</em></td><td style="text-align: right">Nov-2010</td><td style="text-align: right">44,846</td><td style="text-align: right">1,225-bp insertion</td><td style="text-align: right"><code>c</code></td></tr><tr><td style="text-align: right">NZ_CP011656.1</td><td style="text-align: right"><em>C. freundii</em></td><td style="text-align: right">Oct-2012</td><td style="text-align: right">129,196</td><td style="text-align: right">14,960-bp duplication and 70,615-bp insertion</td><td style="text-align: right"><code>e,f,a</code><sup class="footnote-reference"><a id="citeref-2" href="#footnote-2">[2]</a></sup> and <code>g</code></td></tr></table><section class="footnotes is-size-7"><ul><li class="footnote" id="footnote-1"><a class="tag is-link" href="#citeref-1">1</a>your version will probably be different due to the random layout chosen by Bandage. You can always rerun the Bandage command multiple times until you get a nice result.</li><li class="footnote" id="footnote-2"><a class="tag is-link" href="#citeref-2">2</a>The 14,960-bp duplication is represented as a repetition of <code>e,f,a</code> in the bottom plasmid. The total length is <code>5,045+188+9,727=14,960</code> as expected.</li></ul></section></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../tutorial_4/">« Projecting the graph on a subset of strains.</a><a class="docs-footer-nextpage" href="../tutorial_6/">Example application: antibiotic resistance gene neighbourhoods »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 11 February 2025 16:44">Tuesday 11 February 2025</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
