<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Projecting the graph on a subset of strains. · PanGraph.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PanGraph.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorial</span><ul><li><a class="tocitem" href="../tutorial_1/">Building a pangraph</a></li><li><a class="tocitem" href="../tutorial_2/">The structure of Pangraph output file</a></li><li><a class="tocitem" href="../tutorial_3/">Polishing the pangraph and exploring alignments</a></li><li class="is-active"><a class="tocitem" href>Projecting the graph on a subset of strains.</a><ul class="internal"><li><a class="tocitem" href="#Preliminary-steps"><span>Preliminary steps</span></a></li><li><a class="tocitem" href="#Marginalize-the-graph-on-a-subset-of-strains"><span>Marginalize the graph on a subset of strains</span></a></li><li><a class="tocitem" href="#A-look-at-the-marginalized-pangraph"><span>A look at the marginalized pangraph</span></a></li></ul></li><li><a class="tocitem" href="../tutorial_5/">Example application: plasmid rearrangements</a></li><li><a class="tocitem" href="../tutorial_6/">Example application: antibiotic resistance gene neighbourhoods</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/pangraph/">PanGraph globals</a></li><li><a class="tocitem" href="../../lib/align/">Alignment</a></li><li><a class="tocitem" href="../../lib/block/">Blocks</a></li><li><a class="tocitem" href="../../lib/edge/">Edges</a></li><li><a class="tocitem" href="../../lib/graph/">Graphs</a></li><li><a class="tocitem" href="../../lib/gfa/">GFA</a></li><li><a class="tocitem" href="../../lib/mash/">Mash Implementation</a></li><li><a class="tocitem" href="../../lib/minimap/">Minimap2 Wrapper</a></li><li><a class="tocitem" href="../../lib/mmseqs/">MMseqs Wrapper</a></li><li><a class="tocitem" href="../../lib/node/">Nodes</a></li><li><a class="tocitem" href="../../lib/path/">Paths</a></li><li><a class="tocitem" href="../../lib/simulate/">Simulation</a></li><li><a class="tocitem" href="../../lib/utility/">Utility</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/build/">Build</a></li><li><a class="tocitem" href="../../cli/export/">Export</a></li><li><a class="tocitem" href="../../cli/generate/">Generate</a></li><li><a class="tocitem" href="../../cli/marginalize/">Marginalize</a></li><li><a class="tocitem" href="../../cli/polish/">Polish</a></li><li><a class="tocitem" href="../../cli/version/">Version</a></li></ul></li><li><span class="tocitem">Development</span><ul><li><a class="tocitem" href="../../dev/building-docker/">Building Docker image</a></li><li><a class="tocitem" href="../../dev/building-documentation/">Updating the documentation</a></li><li><a class="tocitem" href="../../dev/releasing/">Releasing</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorial</a></li><li class="is-active"><a href>Projecting the graph on a subset of strains.</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Projecting the graph on a subset of strains.</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/neherlab/pangraph/blob/master/docs/src/tutorials/tutorial_4.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Projecting-the-graph-on-a-subset-of-strains."><a class="docs-heading-anchor" href="#Projecting-the-graph-on-a-subset-of-strains.">Projecting the graph on a subset of strains.</a><a id="Projecting-the-graph-on-a-subset-of-strains.-1"></a><a class="docs-heading-anchor-permalink" href="#Projecting-the-graph-on-a-subset-of-strains." title="Permalink"></a></h1><p>In this final part of the tutorial we show how to use the <code>marginalize</code> command. This command is used when one is interested in the relationship between a subset of the genomes that make up the pangraph. It can quickly create, starting from the full pangraph, a smaller pangenome graph relative to the selected subset of strains.</p><h2 id="Preliminary-steps"><a class="docs-heading-anchor" href="#Preliminary-steps">Preliminary steps</a><a id="Preliminary-steps-1"></a><a class="docs-heading-anchor-permalink" href="#Preliminary-steps" title="Permalink"></a></h2><p>We will run this tutorial on a different dataset, containing 10 complete chromosomes of <em>Klebsiella Pneumoniae</em> (source: GenBank). These sequences are available in the pangraph repository (<code>example_dataset/klebs.fa.gz</code>) and can be downloaded by running:</p><pre><code class="language-bash hljs">wget https://github.com/neherlab/pangraph/raw/master/example_datasets/klebs.fa.gz</code></pre><p>As for the previous dataset, we can create the pangraph with the command:</p><pre><code class="language-bash hljs">pangraph build --circular klebs.fa.gz &gt; klebs_pangraph.json</code></pre><p>On 4 cores the command should complete in around 6 mins. After creating the pangraph, we can export it in <code>gfa</code> format for visualization.</p><pre><code class="language-bash hljs">pangraph export \
    --no-duplications \
    --output-directory klebs_export \
    --prefix klebs_pangraph \
    klebs_pangraph.json</code></pre><p>This will create the file <code>klebs_export/klebs_pangraph.gfa</code> which can be visualized using <a href="https://rrwick.github.io/Bandage/">Bandage</a>.</p><p><img src="../../assets/klebs_full_pangraph.png" alt="img"/></p><p>Colors indicate the number of times a block occurs. Blocks that appear in red are core blocks that are found in every chromosome, while black blocks are only present in a few strains. We used the <code>--no-duplication</code> flag in the export command, which excludes duplicated blocks from the exported graph. This simplifies the visualization, which would otherwise be highly &quot;tangled-up&quot; by these duplications.</p><h2 id="Marginalize-the-graph-on-a-subset-of-strains"><a class="docs-heading-anchor" href="#Marginalize-the-graph-on-a-subset-of-strains">Marginalize the graph on a subset of strains</a><a id="Marginalize-the-graph-on-a-subset-of-strains-1"></a><a class="docs-heading-anchor-permalink" href="#Marginalize-the-graph-on-a-subset-of-strains" title="Permalink"></a></h2><p>The full graph might be difficult to interpret. However if we are interested only in relationships between a subset of chromosomes we can use the command <code>marginalize</code> (see <a href="../../cli/marginalize/#Marginalize">Marginalize</a>) to project the pangraph on this set of strains. This &quot;marginalization&quot; operation will remove <em>transitive edges</em>. If two blocks always come one after the other (with same strandedness) in all of the subset of strains one is interested in, these blocks will be merged in a new larger block. This can greatly simplify the pangenome graph, and highlight differences between a particular subset of strains. And it is also computationally much faster than building a new pangraph directly from the sequences of interest.</p><p>For this example we will consider the pair of strains <code>NZ_CP013711</code> and <code>NC_017540</code>.<sup class="footnote-reference"><a id="citeref-1" href="#footnote-1">[1]</a></sup> We can marginalize the pangraph on these two strains by running:</p><pre><code class="language-bash hljs">pangraph marginalize \
    --strains NZ_CP013711,NC_017540 \
    klebs_pangraph.json &gt; klebs_marginal_pangraph.json</code></pre><p>The file <code>klebs_marginal_pangraph.json</code> will contain the new marginalized pangraph. The strains on which one projects are specified with the flag <code>--strains</code>. They must be passed as a comma separated list of sequence ids, without spaces.</p><div class="admonition is-info"><header class="admonition-header">producing all pairwise marginalizations</header><div class="admonition-body"><p>If the command <code>marginalize</code> is executed without specifying particular strains, but providing an output directory with the option <code>--output-path</code>, then it will automatically produce all pairwise marginalizations and save them in the specified directory.</p></div></div><h2 id="A-look-at-the-marginalized-pangraph"><a class="docs-heading-anchor" href="#A-look-at-the-marginalized-pangraph">A look at the marginalized pangraph</a><a id="A-look-at-the-marginalized-pangraph-1"></a><a class="docs-heading-anchor-permalink" href="#A-look-at-the-marginalized-pangraph" title="Permalink"></a></h2><p>As done for the main pangraph, we can export the marginalized pangraph in <code>gfa</code> format:</p><pre><code class="language-bash hljs">pangraph export \
    --no-duplications \
    --output-directory klebs_export \
    --prefix klebs_marginal_pangraph \
    klebs_marginal_pangraph.json</code></pre><p>This will produce the file <code>klebs_export/klebs_marginal_pangraph.gfa</code> which can be visualized with Bandage.</p><p><img src="../../assets/klebs_marginal_pangraph.png" alt="img"/></p><p>As expected the marginalized pangraph contains fewer blocks than the original one (388 vs 1244), and blocks are on average longer (mean length: 14 kbp vs 6 kbp). Blocks that appear in red are shared by both strains, while black blocks are present in only one of the two strains. The pangraph is composed of two stretches of syntenic blocks, which are in contact in a central point. This structure can be understood by comparing the two chromosomes with a dotplot. Using <a href="http://dgenies.toulouse.inra.fr/">D-Genies</a> on the two sequences<sup class="footnote-reference"><a id="citeref-2" href="#footnote-2">[2]</a></sup> we obtain the following:</p><p><img src="../../assets/dgenies_dotplot.png" alt="img"/></p><p>The contact point between the two loops in the pangraph is caused by the fact that the two genomes are composed of two mostly syntenic subsequences (the two loops) but these loops are concatenated with two different strandedness in the two strains. If we were to draw the two paths (relative to the two chromosomes) with different colors on top of the pangraph we would observe something similar to this:</p><p><img src="../../assets/pangraph_inversion.png" alt="img"/></p><section class="footnotes is-size-7"><ul><li class="footnote" id="footnote-1"><a class="tag is-link" href="#citeref-1">1</a>here we consider a marginalization on two strains, but one can marginalize the pangraph on any number of strains.</li><li class="footnote" id="footnote-2"><a class="tag is-link" href="#citeref-2">2</a>to simplify the visualization, the chromosome of strain <code>NZ_CP013711</code> was reverse-complemented and shifted of ~ 0.1 Mbp, so that most of the matches lay on the diagonal.</li></ul></section></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../tutorial_3/">« Polishing the pangraph and exploring alignments</a><a class="docs-footer-nextpage" href="../tutorial_5/">Example application: plasmid rearrangements »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 11 February 2025 15:01">Tuesday 11 February 2025</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
