diff --git a/Makefile b/Makefile
index 4118616..7ba4088 100644
--- a/Makefile
+++ b/Makefile
@@ -53,6 +53,10 @@ minimap2-lite:example.o libminimap2.a
 libminimap2.a:$(OBJS)
 		$(AR) -csru $@ $(OBJS)
 
+libminimap2.SO: CFLAGS += -fpic -I$(INC)
+libminimap2.SO: $(OBJS)
+	$(CC) -shared -o $@ $(OBJS) -lz
+
 sdust:sdust.c kalloc.o kalloc.h kdq.h kvec.h kseq.h ketopt.h sdust.h
 		$(CC) -D_SDUST_MAIN $(CFLAGS) $< kalloc.o -o $@ -lz
 
