<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Edges · PanGraph.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PanGraph.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorial</span><ul><li><a class="tocitem" href="../../tutorials/tutorial_1/">Building a pangraph</a></li><li><a class="tocitem" href="../../tutorials/tutorial_2/">The structure of Pangraph output file</a></li><li><a class="tocitem" href="../../tutorials/tutorial_3/">Polishing the pangraph and exploring alignments</a></li><li><a class="tocitem" href="../../tutorials/tutorial_4/">Projecting the graph on a subset of strains.</a></li><li><a class="tocitem" href="../../tutorials/tutorial_5/">Example application: plasmid rearrangements</a></li><li><a class="tocitem" href="../../tutorials/tutorial_6/">Example application: antibiotic resistance gene neighbourhoods</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../pangraph/">PanGraph globals</a></li><li><a class="tocitem" href="../align/">Alignment</a></li><li><a class="tocitem" href="../block/">Blocks</a></li><li class="is-active"><a class="tocitem" href>Edges</a><ul class="internal"><li><a class="tocitem" href="#Types"><span>Types</span></a></li><li class="toplevel"><a class="tocitem" href="#Functions"><span>Functions</span></a></li></ul></li><li><a class="tocitem" href="../graph/">Graphs</a></li><li><a class="tocitem" href="../gfa/">GFA</a></li><li><a class="tocitem" href="../mash/">Mash Implementation</a></li><li><a class="tocitem" href="../minimap/">Minimap2 Wrapper</a></li><li><a class="tocitem" href="../mmseqs/">MMseqs Wrapper</a></li><li><a class="tocitem" href="../node/">Nodes</a></li><li><a class="tocitem" href="../path/">Paths</a></li><li><a class="tocitem" href="../simulate/">Simulation</a></li><li><a class="tocitem" href="../utility/">Utility</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/build/">Build</a></li><li><a class="tocitem" href="../../cli/export/">Export</a></li><li><a class="tocitem" href="../../cli/generate/">Generate</a></li><li><a class="tocitem" href="../../cli/marginalize/">Marginalize</a></li><li><a class="tocitem" href="../../cli/polish/">Polish</a></li><li><a class="tocitem" href="../../cli/version/">Version</a></li></ul></li><li><span class="tocitem">Development</span><ul><li><a class="tocitem" href="../../dev/building-docker/">Building Docker image</a></li><li><a class="tocitem" href="../../dev/building-documentation/">Updating the documentation</a></li><li><a class="tocitem" href="../../dev/releasing/">Releasing</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>Edges</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Edges</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/neherlab/pangraph/blob/master/docs/src/lib/edge.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Edges"><a class="docs-heading-anchor" href="#Edges">Edges</a><a id="Edges-1"></a><a class="docs-heading-anchor-permalink" href="#Edges" title="Permalink"></a></h1><h2 id="Types"><a class="docs-heading-anchor" href="#Types">Types</a><a id="Types-1"></a><a class="docs-heading-anchor-permalink" href="#Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Edges.Edge" href="#PanGraph.Graphs.Edges.Edge"><code>PanGraph.Graphs.Edges.Edge</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">mutable struct Edge
    block  :: Tuple{Block, Block}
    invert :: Bool # changes strand
    nodes  :: Array{Position}
end</code></pre><p>Store a unique edge within a pangraph. An edge is undirected and is defined by the two juxtaposed blocks, as well as a relative orientation. Contain all positions of all genomes that contain the edge.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/ace1fb2e25fc8a66ca706d32d42ab35a1d56e0bf/src/edge.jl#L142-L152">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Edges.Position" href="#PanGraph.Graphs.Edges.Position"><code>PanGraph.Graphs.Edges.Position</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct Position
    path  :: Path
    node  :: Tuple{Node{Block},Node{Block}}
    index :: Tuple{Int,Int} # positions on path
    locus :: Int # breakpoint on sequence
end</code></pre><p>Store a single position of an edge/breakpoint between homologous pancontigs in an individual genome. <code>path</code> is the containing <code>Path</code> object. <code>node</code> stores the junction of nodes that represent the position of the breakpoint. <code>index</code> is the indices of <code>node</code> within <code>path</code>. <code>locus</code> is the physical location on the genome of the breakpoint.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/ace1fb2e25fc8a66ca706d32d42ab35a1d56e0bf/src/edge.jl#L10-L23">source</a></section></article><h1 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Edges.deparalog!-Tuple{Any}" href="#PanGraph.Graphs.Edges.deparalog!-Tuple{Any}"><code>PanGraph.Graphs.Edges.deparalog!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">deparalog!(G)</code></pre><p>Split duplicated blocks that have non-intersecting, seperable paths that run in parallel and transitively connect associated genomes. Use to simplify high copy number blocks found in all individuals in equivalent contexts within pangraph <code>G</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/ace1fb2e25fc8a66ca706d32d42ab35a1d56e0bf/src/edge.jl#L219-L224">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Edges.edges-Tuple{Any}" href="#PanGraph.Graphs.Edges.edges-Tuple{Any}"><code>PanGraph.Graphs.Edges.edges</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">edges(G)</code></pre><p>Compute all edges associated with pangraph <code>G</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/ace1fb2e25fc8a66ca706d32d42ab35a1d56e0bf/src/edge.jl#L167-L171">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Edges.isolates-Tuple{Vector{PanGraph.Graphs.Edges.Position}}" href="#PanGraph.Graphs.Edges.isolates-Tuple{Vector{PanGraph.Graphs.Edges.Position}}"><code>PanGraph.Graphs.Edges.isolates</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">isolates(positions::Array{Position,1})</code></pre><p>Compute the array of <code>Position</code> values for each isolate.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/ace1fb2e25fc8a66ca706d32d42ab35a1d56e0bf/src/edge.jl#L31-L35">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Edges.next-Tuple{PanGraph.Graphs.Edges.Position, PanGraph.Graphs.Blocks.Block}" href="#PanGraph.Graphs.Edges.next-Tuple{PanGraph.Graphs.Edges.Position, PanGraph.Graphs.Blocks.Block}"><code>PanGraph.Graphs.Edges.next</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">next(x::Position, blk::Block)</code></pre><p>Compute the next position from <code>x</code> that is connected through block <code>blk</code>. Use to traverse the path of an individual genome through a pangraph.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/ace1fb2e25fc8a66ca706d32d42ab35a1d56e0bf/src/edge.jl#L49-L54">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../block/">« Blocks</a><a class="docs-footer-nextpage" href="../graph/">Graphs »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Monday 13 March 2023 14:30">Monday 13 March 2023</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
