<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Alignment · PanGraph.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PanGraph.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorial</span><ul><li><a class="tocitem" href="../../tutorials/tutorial_1/">Building a pangraph</a></li><li><a class="tocitem" href="../../tutorials/tutorial_2/">The structure of Pangraph output file</a></li><li><a class="tocitem" href="../../tutorials/tutorial_3/">Polishing the pangraph and exploring alignments</a></li><li><a class="tocitem" href="../../tutorials/tutorial_4/">Projecting the graph on a subset of strains.</a></li><li><a class="tocitem" href="../../tutorials/tutorial_5/">Example application: plasmid rearrangements</a></li><li><a class="tocitem" href="../../tutorials/tutorial_6/">Example application: antibiotic resistance gene neighbourhoods</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../pangraph/">PanGraph globals</a></li><li class="is-active"><a class="tocitem" href>Alignment</a><ul class="internal"><li><a class="tocitem" href="#Types"><span>Types</span></a></li><li><a class="tocitem" href="#Functions"><span>Functions</span></a></li></ul></li><li><a class="tocitem" href="../block/">Blocks</a></li><li><a class="tocitem" href="../edge/">Edges</a></li><li><a class="tocitem" href="../graph/">Graphs</a></li><li><a class="tocitem" href="../gfa/">GFA</a></li><li><a class="tocitem" href="../mash/">Mash Implementation</a></li><li><a class="tocitem" href="../minimap/">Minimap2 Wrapper</a></li><li><a class="tocitem" href="../mmseqs/">MMseqs Wrapper</a></li><li><a class="tocitem" href="../node/">Nodes</a></li><li><a class="tocitem" href="../path/">Paths</a></li><li><a class="tocitem" href="../simulate/">Simulation</a></li><li><a class="tocitem" href="../utility/">Utility</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/build/">Build</a></li><li><a class="tocitem" href="../../cli/export/">Export</a></li><li><a class="tocitem" href="../../cli/generate/">Generate</a></li><li><a class="tocitem" href="../../cli/marginalize/">Marginalize</a></li><li><a class="tocitem" href="../../cli/polish/">Polish</a></li><li><a class="tocitem" href="../../cli/version/">Version</a></li></ul></li><li><span class="tocitem">Development</span><ul><li><a class="tocitem" href="../../dev/building-docker/">Building Docker image</a></li><li><a class="tocitem" href="../../dev/building-documentation/">Updating the documentation</a></li><li><a class="tocitem" href="../../dev/releasing/">Releasing</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>Alignment</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Alignment</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/neherlab/pangraph/blob/master/docs/src/lib/align.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Alignment"><a class="docs-heading-anchor" href="#Alignment">Alignment</a><a id="Alignment-1"></a><a class="docs-heading-anchor-permalink" href="#Alignment" title="Permalink"></a></h1><h2 id="Types"><a class="docs-heading-anchor" href="#Types">Types</a><a id="Types-1"></a><a class="docs-heading-anchor-permalink" href="#Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.Clade-Tuple{Any, Any}" href="#PanGraph.Graphs.Align.Clade-Tuple{Any, Any}"><code>PanGraph.Graphs.Align.Clade</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Clade(distance, names; algo=:nj)</code></pre><p>Generate a tree from a matrix of pairwise distances <code>distance</code>. The names of leafs are given by an array of strings <code>names</code>. <code>algo</code> dictates the algorithm used to transform the distance matrix into a tree. Currently on neighbor joining (:nj) is supported.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L87-L94">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.Clade-Tuple{Any}" href="#PanGraph.Graphs.Align.Clade-Tuple{Any}"><code>PanGraph.Graphs.Align.Clade</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Clade(name)</code></pre><p>Generate an empty, disconnected clade with name <code>name</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L67-L71">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.Clade-Tuple{PanGraph.Graphs.Align.Clade, PanGraph.Graphs.Align.Clade}" href="#PanGraph.Graphs.Align.Clade-Tuple{PanGraph.Graphs.Align.Clade, PanGraph.Graphs.Align.Clade}"><code>PanGraph.Graphs.Align.Clade</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Clade(left::Clade, right::Clade)</code></pre><p>Generate an nameless clade with <code>left</code> and <code>right</code> children.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L73-L77">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.Clade-Tuple{}" href="#PanGraph.Graphs.Align.Clade-Tuple{}"><code>PanGraph.Graphs.Align.Clade</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Clade()</code></pre><p>Generate an empty, disconnected clade.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L61-L65">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.Message" href="#PanGraph.Graphs.Align.Message"><code>PanGraph.Graphs.Align.Message</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">mutable struct Clade
	name   :: String
	parent :: Union{Clade,Nothing}
	left   :: Union{Clade,Nothing}
	right  :: Union{Clade,Nothing}
	graph  :: Channel{Tuple{Graph,Int}}
end</code></pre><p>Clade is a node (internal or leaf) of a binary guide tree used to order pairwise alignments associated to a multiple genome alignment in progress. <code>name</code> is only non-empty for leaf nodes. <code>parent</code> is <code>nothing</code> for the root node. <code>graph</code> is a 0-sized channel that is used as a message passing primitive in alignment.     It contains the graph and an index used to decide the order of items in a pair in     pairwise graph merge.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L32-L48">source</a></section></article><h2 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.align-Tuple{Function, Vararg{PanGraph.Graphs.Graph}}" href="#PanGraph.Graphs.Align.align-Tuple{Function, Vararg{PanGraph.Graphs.Graph}}"><code>PanGraph.Graphs.Align.align</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">align(aligner::Function, Gs::Graph...; compare=Mash.distance, energy=(hit)-&gt;(-Inf), minblock=100, reference=nothing, maxiter=100)</code></pre><p>Aligns a collection of graphs <code>Gs</code> using the specified <code>aligner</code> function to recover hits. Graphs are aligned following an internal guide tree, generated using kmer distance.</p><p><code>energy</code> is to be a function that takes an alignment between two blocks and produces a score. The <em>lower</em> the score, the <em>better</em> the alignment. Only negative energies are considered.</p><p><code>minblock</code> is the minimum size block that will be produced from the algorithm. <code>maxiter</code> is maximum number of duplications that will be considered during this alignment.</p><p><code>compare</code> is the function to be used to generate pairwise distances that generate the internal guide tree.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L620-L633">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.align_pair-Tuple{PanGraph.Graphs.Graph, PanGraph.Graphs.Graph, Function, Int64, Function, Function, Bool}" href="#PanGraph.Graphs.Align.align_pair-Tuple{PanGraph.Graphs.Graph, PanGraph.Graphs.Graph, Function, Int64, Function, Function, Bool}"><code>PanGraph.Graphs.Align.align_pair</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">align_pair(G₁::Graph, G₂::Graph, energy::Function, minblock::Int, verify::Function, verbose::Bool; maxiter=100)</code></pre><p>Align graph <code>G₁</code> to graph <code>G₂</code> by looking for homology between consensus sequences of blocks. This is a low-level function.</p><p><code>energy</code> is to be a function that takes an alignment between two blocks and produces a score. The <em>lower</em> the score, the <em>better</em> the alignment. Only negative energies are considered.</p><p><code>minblock</code> is the minimum size block that will be produced from the algorithm. <code>maxiter</code> is maximum number of duplications that will be considered during this alignment.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L558-L570">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.align_self-Tuple{PanGraph.Graphs.Graph, Function, Int64, Function, Function, Bool}" href="#PanGraph.Graphs.Align.align_self-Tuple{PanGraph.Graphs.Graph, Function, Int64, Function, Function, Bool}"><code>PanGraph.Graphs.Align.align_self</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">align_self(G₁::Graph, energy::Function, minblock::Int, verify::Function, verbose::Bool; maxiter=100)</code></pre><p>Align graph <code>G₁</code> to itself by looking for homology between blocks. This is a low-level function.</p><p><code>energy</code> is to be a function that takes an alignment between two blocks and produces a score. The <em>lower</em> the score, the <em>better</em> the alignment. Only negative energies are considered.</p><p><code>minblock</code> is the minimum size block that will be produced from the algorithm. <code>maxiter</code> is maximum number of duplications that will be considered during this alignment.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L422-L433">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.balance-Tuple{PanGraph.Graphs.Align.Clade}" href="#PanGraph.Graphs.Align.balance-Tuple{PanGraph.Graphs.Align.Clade}"><code>PanGraph.Graphs.Align.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">balance(root::Clade)</code></pre><p>Rotate a binary tree to a balanced configuration. Preserves the topological ordering of the leaves.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L288-L293">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.isleaf-Tuple{PanGraph.Graphs.Align.Clade}" href="#PanGraph.Graphs.Align.isleaf-Tuple{PanGraph.Graphs.Align.Clade}"><code>PanGraph.Graphs.Align.isleaf</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">isleaf(c::Clade)</code></pre><p>Return if Clade <code>c</code> is a terminal node, i.e. a leaf.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L184-L188">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.leaves-Tuple{PanGraph.Graphs.Align.Clade}" href="#PanGraph.Graphs.Align.leaves-Tuple{PanGraph.Graphs.Align.Clade}"><code>PanGraph.Graphs.Align.leaves</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">leaves(root::Clade)</code></pre><p>Return all terminal nodes that have <code>root</code> as their common ancestor.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L204-L208">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.nj-Tuple{Any, Any}" href="#PanGraph.Graphs.Align.nj-Tuple{Any, Any}"><code>PanGraph.Graphs.Align.nj</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">nj(distance, names)</code></pre><p>Lower-level function. Generate a tree from a matrix of pairwise distances <code>distance</code>. The names of leafs are given by an array of strings <code>names</code>. Uses neighbor joining.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L102-L109">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.ordering-Tuple{Any, Vararg{Any}}" href="#PanGraph.Graphs.Align.ordering-Tuple{Any, Vararg{Any}}"><code>PanGraph.Graphs.Align.ordering</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ordering(compare, Gs...)</code></pre><p>Return a guide tree based upon distances computed from a collection of graphs <code>Gs</code>, using method <code>compare</code>. The signature of <code>compare</code> is expected to be <code>compare(G::Graphs....) -&gt; distance, names</code>. Graphs <code>Gs...</code> are expected to be singleton graphs.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L313-L319">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.postorder-Tuple{PanGraph.Graphs.Align.Clade}" href="#PanGraph.Graphs.Align.postorder-Tuple{PanGraph.Graphs.Align.Clade}"><code>PanGraph.Graphs.Align.postorder</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">postorder(root::Clade)</code></pre><p>Return an postorder iterator over descendents of <code>root</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L228-L232">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PanGraph.Graphs.Align.preorder-Tuple{PanGraph.Graphs.Align.Clade}" href="#PanGraph.Graphs.Align.preorder-Tuple{PanGraph.Graphs.Align.Clade}"><code>PanGraph.Graphs.Align.preorder</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">preorder(root::Clade)</code></pre><p>Return an preorder iterator over descendents of <code>root</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/neherlab/pangraph/blob/062e1221e4ef6e003d726be9a4ecc05a6bf1bdfc/src/align.jl#L253-L257">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../pangraph/">« PanGraph globals</a><a class="docs-footer-nextpage" href="../block/">Blocks »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Wednesday 12 July 2023 14:52">Wednesday 12 July 2023</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
